<!DOCTYPE html>
<html><head>
<meta http-equiv="content-type" content="text/html; charset=UTF-8">
        <meta charset="utf-8">
        <meta http-equiv="content-language" content="en">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <title> - ENCODE</title>
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <link rel="stylesheet" href="SearchENCODE_files/style.css">
        <link rel="stylesheet" href="SearchENCODE_files/responsive.css">
        <script data-main="/static/main" src="SearchENCODE_files/require.js"></script>
    <script src="SearchENCODE_files/main.js" data-requiremodule="main" data-requirecontext="_" async="" charset="utf-8" type="text/javascript"></script><script src="SearchENCODE_files/jquery.js" data-requiremodule="jquery" data-requirecontext="_" async="" charset="utf-8" type="text/javascript"></script><script src="SearchENCODE_files/jsx.js" data-requiremodule="jsx" data-requirecontext="_" async="" charset="utf-8" type="text/javascript"></script><script src="SearchENCODE_files/react.js" data-requiremodule="react" data-requirecontext="_" async="" charset="utf-8" type="text/javascript"></script><script src="SearchENCODE_files/include.js" data-requiremodule="persona" data-requirecontext="_" async="" charset="utf-8" type="text/javascript"></script><script src="SearchENCODE_files/underscore.js" data-requiremodule="underscore" data-requirecontext="_" async="" charset="utf-8" type="text/javascript"></script><script src="SearchENCODE_files/bootstrap.js" data-requiremodule="bootstrap" data-requirecontext="_" async="" charset="utf-8" type="text/javascript"></script><script src="SearchENCODE_files/sticky_header.js" data-requiremodule="stickyheader" data-requirecontext="_" async="" charset="utf-8" type="text/javascript"></script><script src="SearchENCODE_files/jsonform.js" data-requiremodule="jsonform" data-requirecontext="_" async="" charset="utf-8" type="text/javascript"></script><script src="SearchENCODE_files/JSXTransformer.js" data-requiremodule="JSXTransformer" data-requirecontext="_" async="" charset="utf-8" type="text/javascript"></script><script src="SearchENCODE_files/globals.js" data-requiremodule="globals" data-requirecontext="_" async="" charset="utf-8" type="text/javascript"></script><script src="SearchENCODE_files/uri.js" data-requiremodule="uri" data-requirecontext="_" async="" charset="utf-8" type="text/javascript"></script><script src="SearchENCODE_files/mixins.js" data-requiremodule="mixins" data-requirecontext="_" async="" charset="utf-8" type="text/javascript"></script><script src="SearchENCODE_files/class.js" data-requiremodule="class" data-requirecontext="_" async="" charset="utf-8" type="text/javascript"></script><script src="SearchENCODE_files/registry.js" data-requiremodule="registry" data-requirecontext="_" async="" charset="utf-8" type="text/javascript"></script></head>
    <body>
        <script id="data-context" type="application/ld+json">{"portal_title": "ENCODE 3", "title": "ElasticSearch View", "items": {"antibodies": [{"antibody_lot_uuid": "480e7b58-5b3b-4614-80c6-f57c7958d0db", "target_uuid": "96e54052-bca1-4138-b004-d50beaeb24f6", "approval_status": "FAILED", "validation_uuids": ["89caf127-b3a7-40ac-b50d-10b341613ba0"], "validations": [{"lab_uuid": "17e5f9cf-d429-4794-87a6-cac083e3b275", "document": {"download": "hnRNPLL_aviva-2_IP-WB_HeLa_Fu.TIF", "width": 3072, "href": "@@download/document/hnRNPLL_aviva-2_IP-WB_HeLa_Fu.TIF", "type": "image/tiff", "height": 2304}, "target_uuid": "96e54052-bca1-4138-b004-d50beaeb24f6", "product_id": "ARP41103_P050", "antibody_lot": {"source_uuid": "fa701215-07e6-4ffe-a8f4-20356d66f3e0", "lab_uuid": "17e5f9cf-d429-4794-87a6-cac083e3b275", "antigen_sequence": "MSSSSSSPRETYEEDREYESQAKRLKTEEGEIDYSAEEGENRREATPRGG", "submitter_uuid": "cc0edf23-b9fd-48e2-8609-54ad07bdd5f7", "target_label": "HNRPLL", "product_id": "ARP41103_P050", "isotype": "IgG", "clonality": "polyclonal", "award_uuid": "d85c9929-1169-472e-b377-3f7dbbe61e09", "project": "ENCODE3", "source": {"source_name": "Aviva Systems Biology", "url": "http://www.avivasysbio.com", "actions": [{"profile": "/profiles/source.json", "href": "", "method": "POST", "name": "edit", "title": "Edit"}], "alias": "Aviva", "test": "test", "@id": "/sources/fa701215-07e6-4ffe-a8f4-20356d66f3e0/", "@type": ["source", "item"]}, "antibody_accession": "ENCAB000AWB", "url": "http://www.avivasysbio.com/anti-hnrpll-antibody-n-terminal-region-arp41103-p050.html", "host_organism": "rabbit", "lot_id": "QC21844", "dbxref": {"encode2": []}, "@id": "/antibody-lots/480e7b58-5b3b-4614-80c6-f57c7958d0db/", "@type": ["antibody_lot", "item"]}, "antibody_lot_uuid": "480e7b58-5b3b-4614-80c6-f57c7958d0db", "validation_status": "FAILED", "award_uuid": "d85c9929-1169-472e-b377-3f7dbbe61e09", "validation_method": "Western blot after IP", "project": "ENCODE3", "caption": "Immunoprecipitation from HeLa whole cell lysate and analized by western blot analysis uisng rabbit polyclonal to HNRPLL", "is_current": "TRUE", "award": {"award_name": null, "description": "The goal of this proposal is to comprehensively characterize the functional sequence elements encoded in the human genome that are recognized by 250 RNA binding proteins (RBPs) in two cell lines. To do this, we will generate stable HeLa-S3 and GM 12878 cell lines expressing epitope-tagged RBPs and determine the sub cellular localization pattern of each RBP. These cells will be used to perform CLIP-Seq assays to define genome-wide, and at single-nucleotide resolution, the RNA sequence elements recognized by 250 RBPs. The RNA sequence elements identified will be validated using sequence-based in vitro binding assays. Furthermore, ChlP-Seq will be performed for all nuclear localized RBPs to determine the regions of the genome and chromatin that each RBP associates with. These binding assays will be supplemented with functional assays in RBP-depleted cells that will be critical for assigning functions to the identified binding sites. These assays include RNA-Seq of total cellular RNA and RNA purified from various cellular fractions, ribosomal footprint profiling, and Gro-Seq. Together, these assays will provide functional information regarding the roles of each RBP in splicing, cleavage and polyadenylation, RNA stability, RNA editing, translation, RNA localization, and transcription. Bioinformatic analysis will be performed, largely using software generated by our group, to quantitate all assays and to associate functions to the sequence elements identified in the binding assays. Together, these experiments will provide a comprehensive and in depth measure of the functions of approximately half of the human RBPs and the functional sequence elements that they interact with. This project will fill a major gap in the catalog of functional elements encoded in the human genome that are being characterized by the ENCODE consortium. The product of this project will be a unique and valuable community resource that will push the field forward in new and exciting ways and will almost certainly create new paradigms regarding the functions of RBPs and RNA-protein networks in human biology and disease.", "end_date": "2016-0731", "pi_first_name": "Brenton R", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402439", "award_end_date": null, "award_start_date": null, "number": "U54HG007005", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Graveley", "@id": "/awards/d85c9929-1169-472e-b377-3f7dbbe61e09/", "@type": ["award", "item"], "name": "COMPRHENSIVE ANALYSIS OF FUNCTIONAL RNA ELEMENTS ENCODED IN THE HUMAN GENOME"}, "lab": {"award_number_list": [], "fax": "858-534-8549", "institute_name": "University of California at San Diego", "name": "Xiang-Dong Fu, UCSD", "city": "La Jolla", "phone2": null, "address1": "Department of Cellular and Molecular Medicine", "address2": "9500 Gilman Drive, Room 231", "phone1": "858-534-4937", "award_uuids": [], "url": null, "state": "CA", "postal_code": "92093-0651", "institute_label": "UCSD", "test": "test", "awards": [], "pi_name": null, "country": "USA", "@id": "/labs/17e5f9cf-d429-4794-87a6-cac083e3b275/", "@type": ["lab", "item"], "lab_name": null}, "lot_id": "QC21844", "submitter_uuid": "cc0edf23-b9fd-48e2-8609-54ad07bdd5f7", "@id": "/validations/89caf127-b3a7-40ac-b50d-10b341613ba0/", "@type": ["antibody_validation", "item"], "submitter": {"lab_uuids": ["17e5f9cf-d429-4794-87a6-cac083e3b275"], "fax": "Unknown", "last_name": "Tsutsui", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "858-534-4926", "first_name": "Taiki", "@type": ["user", "item"], "labs": ["/labs/17e5f9cf-d429-4794-87a6-cac083e3b275"], "lab_name_list": ["Xiang-Dong Fu, UCSD"], "skype": "Unknown", "test": "test", "timezone": "UTC-8", "@id": "/users/cc0edf23-b9fd-48e2-8609-54ad07bdd5f7/", "email": "ttsutsui@ucsd.edu", "job_title": "Post-doc"}, "target": {"lab_uuid": "17e5f9cf-d429-4794-87a6-cac083e3b275", "date_created": "2013-04-08", "submitter_uuid": "cc0edf23-b9fd-48e2-8609-54ad07bdd5f7", "target_label": "HNRPLL", "uniprot_dbxref_list": "Q8WVV9", "@id": "/targets/96e54052-bca1-4138-b004-d50beaeb24f6/", "award_uuid": "d85c9929-1169-472e-b377-3f7dbbe61e09", "award": {"award_name": null, "description": "The goal of this proposal is to comprehensively characterize the functional sequence elements encoded in the human genome that are recognized by 250 RNA binding proteins (RBPs) in two cell lines. To do this, we will generate stable HeLa-S3 and GM 12878 cell lines expressing epitope-tagged RBPs and determine the sub cellular localization pattern of each RBP. These cells will be used to perform CLIP-Seq assays to define genome-wide, and at single-nucleotide resolution, the RNA sequence elements recognized by 250 RBPs. The RNA sequence elements identified will be validated using sequence-based in vitro binding assays. Furthermore, ChlP-Seq will be performed for all nuclear localized RBPs to determine the regions of the genome and chromatin that each RBP associates with. These binding assays will be supplemented with functional assays in RBP-depleted cells that will be critical for assigning functions to the identified binding sites. These assays include RNA-Seq of total cellular RNA and RNA purified from various cellular fractions, ribosomal footprint profiling, and Gro-Seq. Together, these assays will provide functional information regarding the roles of each RBP in splicing, cleavage and polyadenylation, RNA stability, RNA editing, translation, RNA localization, and transcription. Bioinformatic analysis will be performed, largely using software generated by our group, to quantitate all assays and to associate functions to the sequence elements identified in the binding assays. Together, these experiments will provide a comprehensive and in depth measure of the functions of approximately half of the human RBPs and the functional sequence elements that they interact with. This project will fill a major gap in the catalog of functional elements encoded in the human genome that are being characterized by the ENCODE consortium. The product of this project will be a unique and valuable community resource that will push the field forward in new and exciting ways and will almost certainly create new paradigms regarding the functions of RBPs and RNA-protein networks in human biology and disease.", "end_date": "2016-0731", "pi_first_name": "Brenton R", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402439", "award_end_date": null, "award_start_date": null, "number": "U54HG007005", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Graveley", "@id": "/awards/d85c9929-1169-472e-b377-3f7dbbe61e09/", "@type": ["award", "item"], "name": "COMPRHENSIVE ANALYSIS OF FUNCTIONAL RNA ELEMENTS ENCODED IN THE HUMAN GENOME"}, "project": "ENCODE3", "target_gene_name": "HNRPLL", "geneid_dbxref_list": 92906, "is_current": true, "lab": {"award_number_list": [], "fax": "858-534-8549", "institute_name": "University of California at San Diego", "name": "Xiang-Dong Fu, UCSD", "city": "La Jolla", "phone2": null, "address1": "Department of Cellular and Molecular Medicine", "address2": "9500 Gilman Drive, Room 231", "phone1": "858-534-4937", "award_uuids": [], "url": null, "state": "CA", "postal_code": "92093-0651", "institute_label": "UCSD", "test": "test", "awards": [], "pi_name": null, "country": "USA", "@id": "/labs/17e5f9cf-d429-4794-87a6-cac083e3b275/", "@type": ["lab", "item"], "lab_name": null}, "organism_uuid": "7745b647-ff15-4ff3-9ced-b897d4e2983c", "submitter": {"lab_uuids": ["17e5f9cf-d429-4794-87a6-cac083e3b275"], "fax": "Unknown", "last_name": "Tsutsui", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "858-534-4926", "first_name": "Taiki", "@type": ["user", "item"], "labs": ["/labs/17e5f9cf-d429-4794-87a6-cac083e3b275"], "lab_name_list": ["Xiang-Dong Fu, UCSD"], "skype": "Unknown", "test": "test", "timezone": "UTC-8", "@id": "/users/cc0edf23-b9fd-48e2-8609-54ad07bdd5f7/", "email": "ttsutsui@ucsd.edu", "job_title": "Post-doc"}, "ensembl_dbxref": "ENSG00000143889", "dbxref": {"geneid": ["92906"], "uniprot": ["Q8WVV9"], "hgnc": []}, "organism": {"scientific_name": "Homo sapiens", "taxon_id": 9606, "organism_name": "human", "@id": "/organisms/7745b647-ff15-4ff3-9ced-b897d4e2983c/", "@type": ["organism", "item"]}, "@type": ["target", "item"]}}], "project": "ENCODE3", "antibody_lot": {"source_uuid": "fa701215-07e6-4ffe-a8f4-20356d66f3e0", "lab_uuid": "17e5f9cf-d429-4794-87a6-cac083e3b275", "antigen_sequence": "MSSSSSSPRETYEEDREYESQAKRLKTEEGEIDYSAEEGENRREATPRGG", "submitter_uuid": "cc0edf23-b9fd-48e2-8609-54ad07bdd5f7", "target_label": "HNRPLL", "product_id": "ARP41103_P050", "isotype": "IgG", "clonality": "polyclonal", "award_uuid": "d85c9929-1169-472e-b377-3f7dbbe61e09", "project": "ENCODE3", "source": {"source_name": "Aviva Systems Biology", "url": "http://www.avivasysbio.com", "actions": [{"profile": "/profiles/source.json", "href": "", "method": "POST", "name": "edit", "title": "Edit"}], "alias": "Aviva", "test": "test", "@id": "/sources/fa701215-07e6-4ffe-a8f4-20356d66f3e0/", "@type": ["source", "item"]}, "antibody_accession": "ENCAB000AWB", "url": "http://www.avivasysbio.com/anti-hnrpll-antibody-n-terminal-region-arp41103-p050.html", "host_organism": "rabbit", "lot_id": "QC21844", "dbxref": {"encode2": []}, "@id": "/antibody-lots/480e7b58-5b3b-4614-80c6-f57c7958d0db/", "@type": ["antibody_lot", "item"]}, "@id": "/antibodies/25d9b9e8-9ca9-4397-9fb6-e4c698c9cfb8/", "@type": ["antibody_approval", "item"], "target": {"lab_uuid": "17e5f9cf-d429-4794-87a6-cac083e3b275", "date_created": "2013-04-08", "submitter_uuid": "cc0edf23-b9fd-48e2-8609-54ad07bdd5f7", "target_label": "HNRPLL", "uniprot_dbxref_list": "Q8WVV9", "@id": "/targets/96e54052-bca1-4138-b004-d50beaeb24f6/", "award_uuid": "d85c9929-1169-472e-b377-3f7dbbe61e09", "award": {"award_name": null, "description": "The goal of this proposal is to comprehensively characterize the functional sequence elements encoded in the human genome that are recognized by 250 RNA binding proteins (RBPs) in two cell lines. To do this, we will generate stable HeLa-S3 and GM 12878 cell lines expressing epitope-tagged RBPs and determine the sub cellular localization pattern of each RBP. These cells will be used to perform CLIP-Seq assays to define genome-wide, and at single-nucleotide resolution, the RNA sequence elements recognized by 250 RBPs. The RNA sequence elements identified will be validated using sequence-based in vitro binding assays. Furthermore, ChlP-Seq will be performed for all nuclear localized RBPs to determine the regions of the genome and chromatin that each RBP associates with. These binding assays will be supplemented with functional assays in RBP-depleted cells that will be critical for assigning functions to the identified binding sites. These assays include RNA-Seq of total cellular RNA and RNA purified from various cellular fractions, ribosomal footprint profiling, and Gro-Seq. Together, these assays will provide functional information regarding the roles of each RBP in splicing, cleavage and polyadenylation, RNA stability, RNA editing, translation, RNA localization, and transcription. Bioinformatic analysis will be performed, largely using software generated by our group, to quantitate all assays and to associate functions to the sequence elements identified in the binding assays. Together, these experiments will provide a comprehensive and in depth measure of the functions of approximately half of the human RBPs and the functional sequence elements that they interact with. This project will fill a major gap in the catalog of functional elements encoded in the human genome that are being characterized by the ENCODE consortium. The product of this project will be a unique and valuable community resource that will push the field forward in new and exciting ways and will almost certainly create new paradigms regarding the functions of RBPs and RNA-protein networks in human biology and disease.", "end_date": "2016-0731", "pi_first_name": "Brenton R", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402439", "award_end_date": null, "award_start_date": null, "number": "U54HG007005", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Graveley", "@id": "/awards/d85c9929-1169-472e-b377-3f7dbbe61e09/", "@type": ["award", "item"], "name": "COMPRHENSIVE ANALYSIS OF FUNCTIONAL RNA ELEMENTS ENCODED IN THE HUMAN GENOME"}, "project": "ENCODE3", "target_gene_name": "HNRPLL", "geneid_dbxref_list": 92906, "is_current": true, "lab": {"award_number_list": [], "fax": "858-534-8549", "institute_name": "University of California at San Diego", "name": "Xiang-Dong Fu, UCSD", "city": "La Jolla", "phone2": null, "address1": "Department of Cellular and Molecular Medicine", "address2": "9500 Gilman Drive, Room 231", "phone1": "858-534-4937", "award_uuids": [], "url": null, "state": "CA", "postal_code": "92093-0651", "institute_label": "UCSD", "test": "test", "awards": [], "pi_name": null, "country": "USA", "@id": "/labs/17e5f9cf-d429-4794-87a6-cac083e3b275/", "@type": ["lab", "item"], "lab_name": null}, "organism_uuid": "7745b647-ff15-4ff3-9ced-b897d4e2983c", "submitter": {"lab_uuids": ["17e5f9cf-d429-4794-87a6-cac083e3b275"], "fax": "Unknown", "last_name": "Tsutsui", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "858-534-4926", "first_name": "Taiki", "@type": ["user", "item"], "labs": ["/labs/17e5f9cf-d429-4794-87a6-cac083e3b275"], "lab_name_list": ["Xiang-Dong Fu, UCSD"], "skype": "Unknown", "test": "test", "timezone": "UTC-8", "@id": "/users/cc0edf23-b9fd-48e2-8609-54ad07bdd5f7/", "email": "ttsutsui@ucsd.edu", "job_title": "Post-doc"}, "ensembl_dbxref": "ENSG00000143889", "dbxref": {"geneid": ["92906"], "uniprot": ["Q8WVV9"], "hgnc": []}, "organism": {"scientific_name": "Homo sapiens", "taxon_id": 9606, "organism_name": "human", "@id": "/organisms/7745b647-ff15-4ff3-9ced-b897d4e2983c/", "@type": ["organism", "item"]}, "@type": ["target", "item"]}}, {"antibody_lot_uuid": "b98aaa1c-f594-4059-aefd-621f72a57f9e", "target_uuid": "0369858d-7f87-42c4-84ff-7a352e825ffa", "approval_status": "APPROVED", "validation_uuids": ["74f11c08-b865-4de1-88bb-462165f390af", "612e938b-ccbb-4e68-894f-8702edc9181f"], "validations": [{"lab_uuid": "72a9c49a-4e17-4df1-96d1-fd19aa46b175", "document": {"download": "human_CtBP2_validation_Farnham.pdf", "href": "@@download/document/human_CtBP2_validation_Farnham.pdf", "type": "application/pdf"}, "target_uuid": "0369858d-7f87-42c4-84ff-7a352e825ffa", "date_validated": "Unknown", "antibody_lot_uuid": "b98aaa1c-f594-4059-aefd-621f72a57f9e", "validation_status": "APPROVED", "award_uuid": "df972196-c3c7-4a58-a852-94baa87f9b71", "validation_method": "Western blot", "project": "ENCODE2", "caption": "Western blot analysis of Ntera2 nuclear extract.", "lab": {"award_number_list": ["RC2HG005602", "U54HG004558", "U54HG006996"], "fax": "650-725-1534", "institute_name": "Stanford University", "name": "Michael Snyder, Stanford", "city": "Stanford", "phone2": "650-736-8099", "address1": "Department of Genetics", "address2": "300 Pasteur Drive; MC5120", "phone1": "650-723-4668", "award_uuids": ["95839ec3-9b83-44d9-a98a-026a9fd01bda", "df972196-c3c7-4a58-a852-94baa87f9b71", "4d462953-2da5-4fcf-a695-7206f2d5cf45"], "url": null, "state": "CA", "postal_code": "94305-5120", "institute_label": "Stanford", "test": "test", "awards": ["/awards/95839ec3-9b83-44d9-a98a-026a9fd01bda", "/awards/df972196-c3c7-4a58-a852-94baa87f9b71", "/awards/4d462953-2da5-4fcf-a695-7206f2d5cf45"], "pi_name": null, "country": "USA", "@id": "/labs/72a9c49a-4e17-4df1-96d1-fd19aa46b175/", "@type": ["lab", "item"], "lab_name": null}, "award": {"award_name": null, "description": "Mapping regulatory information throughout the human genome is essential for both determining the function of the human genome and for understanding human development and disease. We propose to establish a high throughput production facility to identify the DNA binding sites for transcription factors and chromatin modifying proteins in human cells. Transcription factor targets, chromatin modifications and modifying enzymes will be mapped across the genome using chromatin immunoprecipitation experiments. This binding information will be validated and analyzed to ensure that it is of high quality and then compared to other genomic annotations. Our three aims are: 1) To establish a production facility for mapping targets of transcription factors. The pipeline will consist of five components: a) cell preparation, b) antibody validation and chromatin immunoprecipitation, c) DNA target analysis, d) target validation and e) sample tracking and databases. 2) To use the pipeline to map targets of 600 transcription factors, chromatin modifications and chromatin modifying enzymes in at least two major cell lines (HeLa S3 and HepG2), and targets of other factors in additional cell lines. 3) To maintain our results in a database and also make them accessible through public databases. We will interrelate our genomic annotations with those determined by other groups (particularly those from the ENCODE consortium), focusing on correlating patterns of transcription-factor binding and chromatin modifications with gene expression. We expect this project to contribute significant information to the identification of regulatory and functional elements in the human genome.", "end_date": "2013-0630", "pi_first_name": "Michael P", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8325805", "award_end_date": null, "award_start_date": null, "number": "U54HG004558", "project": "ENCODE2", "award_number": null, "start_date": "2007-0928", "test": "test", "award_url": null, "pi_last_name": "Snyder", "@id": "/awards/df972196-c3c7-4a58-a852-94baa87f9b71/", "@type": ["award", "item"], "name": "PRODUCTION CENTER FOR GLOBAL MAPPING OF REGULATORY ELEMENTS"}, "target": {"lab_uuid": "fb0af3d0-3a4c-4e96-b67a-f273fe527b04", "date_created": "2012-09-30", "hgnc_dbxref_list": "ribeye", "target_label": "CTBP2", "uniprot_dbxref_list": "P56545", "@id": "/targets/0369858d-7f87-42c4-84ff-7a352e825ffa/", "award_uuid": "1a4d6443-8e29-4b4a-99dd-f93e72d42418", "award": {"award_name": null, "description": "We propose establishing the UCSC ENCODE Data Coordination Center to collect, organize, store, and provide access to data from the ENCODE project and other related projects, based on the well-established UCSC Genome Browser. The UCSC browser is a robust, flexible platform that integrates a variety of features that make accessing genomic data both convenient and information-rich. As such, it has attained a high level of acceptance throughout the biology, bioinformatics, and biomedical research communities. During the pilot phase of the ENCODE project, involving 1% of the genome, UCSC has been the repository of sequence-based data. Through participation in the pilot phase, we have established cost-effective, high-throughput approaches for incorporating and displaying ENCODE data, and we have developed an effective interface to the consortium for uploading data files and methods documentation to our FTP site. Through ongoing expansion and improvement of our ENCODE browser, we have laid the groundwork for data management in the whole-genome phase of the project. UCSC's specific aims for participation as the ENCODE DCC are as follows: 1) coordinate closely with the ENCODE data providers to collect, store, and manage ENCODE sequence-based functional element data and related metadata; 2) provide and maintain a long-term storage system for data from the ENCODE Consortium and related genomic projects; 3) provide the ENCODE and public research communities with freely-available mechanisms to access, search, and analyze the ENCODE data; 4) employ good software engineering practices to ensure that the ENCODE data are of the highest possible quality; and 5) foster collaboration with the biomedical community. Fundamental to understanding human biology and disease is the ability to identify and locate the functional elements in the human genome sequence. Both the effective presentation of experimental data and the ability to analyze sequence data using computational methods will help elucidate genomic function. Because the UCSC Genome Browser serves both of these purposes, it offers the ideal platform for ENCODE data.", "end_date": "2012-1231", "pi_first_name": "W James", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8332919", "award_end_date": null, "award_start_date": null, "number": "U41HG004568", "project": "ENCODE2", "award_number": null, "start_date": "2007-0928", "test": "test", "award_url": null, "pi_last_name": "Kent", "@id": "/awards/1a4d6443-8e29-4b4a-99dd-f93e72d42418/", "@type": ["award", "item"], "name": "THE UCSC ENCODE DATA COORDINATION CENTER"}, "project": "ENCODE2", "target_gene_name": "CTBP2", "geneid_dbxref_list": 1488, "is_current": true, "lab": {"award_number_list": ["U41HG004568"], "fax": "Unknown", "institute_name": "University of California at Santa Cruz", "name": "Jim Kent, UCSC", "city": "Santa Cruz", "phone2": null, "address1": "Baskin School of Engineering", "address2": "1156 High St.; Mail Stop CBSE/ITI", "phone1": "831-459-1401", "award_uuids": ["1a4d6443-8e29-4b4a-99dd-f93e72d42418"], "url": null, "state": "CA", "postal_code": "95064", "institute_label": "UCSC", "test": "test", "awards": ["/awards/1a4d6443-8e29-4b4a-99dd-f93e72d42418"], "pi_name": null, "country": "USA", "@id": "/labs/fb0af3d0-3a4c-4e96-b67a-f273fe527b04/", "@type": ["lab", "item"], "lab_name": null}, "organism_uuid": "7745b647-ff15-4ff3-9ced-b897d4e2983c", "submitter": {"lab_uuids": ["fb0af3d0-3a4c-4e96-b67a-f273fe527b04", "cfb789b8-46f3-4d59-a2b3-adc39e7df93a"], "fax": "Unknown", "last_name": "Kent", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "831-459-1401", "first_name": "Jim", "@type": ["user", "item"], "labs": ["/labs/fb0af3d0-3a4c-4e96-b67a-f273fe527b04", "/labs/cfb789b8-46f3-4d59-a2b3-adc39e7df93a"], "lab_name_list": ["Jim Kent, UCSC", "Michael Cherry, Stanford"], "skype": "Unknown", "test": "test", "timezone": "UTC-8", "@id": "/users/746eef27-d857-4b38-a469-cac93fb02164/", "email": "kent@soe.ucsc.edu", "job_title": "PI"}, "ensembl_dbxref": "ENSG00000175029", "target_class": "Unknown", "submitter_uuid": "746eef27-d857-4b38-a469-cac93fb02164", "organism": {"scientific_name": "Homo sapiens", "taxon_id": 9606, "organism_name": "human", "@id": "/organisms/7745b647-ff15-4ff3-9ced-b897d4e2983c/", "@type": ["organism", "item"]}, "@type": ["target", "item"], "dbxref": {"geneid": ["1488"], "uniprot": ["P56545"], "hgnc": ["ribeye"]}}, "antibody_lot": {"source_uuid": "409ad2e5-845f-4203-9440-0a55298828a9", "lab_uuid": "72a9c49a-4e17-4df1-96d1-fd19aa46b175", "encode2_dbxref_list": "CtBP2", "dbxref": {"encode2": ["CtBP2"]}, "target_label": "CTBP2", "product_id": "612044", "isotype": "Unknown", "clonality": "monoclonal", "award_uuid": "df972196-c3c7-4a58-a852-94baa87f9b71", "purification": "Unknown", "project": "ENCODE2", "source": {"source_name": "BD Biosciences", "url": "http://www.bdbiosciences.com", "actions": [{"profile": "/profiles/source.json", "href": "", "method": "POST", "name": "edit", "title": "Edit"}], "alias": "BD Biosciences", "test": "test", "@id": "/sources/409ad2e5-845f-4203-9440-0a55298828a9/", "@type": ["source", "item"]}, "antigen_description": "Unknown", "antibody_accession": "ENCAB000AFO", "url": "http://www.bdbiosciences.com/ptProduct.jsp?prodId=229326", "antigen_sequence": "Unknown", "host_organism": "mouse", "lot_id": "Unknown", "submitter_uuid": "2eb068c5-b7a6-48ec-aca2-c439e4dabb08", "@id": "/antibody-lots/b98aaa1c-f594-4059-aefd-621f72a57f9e/", "@type": ["antibody_lot", "item"]}, "lot_id": "Unknown", "submitter_uuid": "2eb068c5-b7a6-48ec-aca2-c439e4dabb08", "@id": "/validations/74f11c08-b865-4de1-88bb-462165f390af/", "@type": ["antibody_validation", "item"], "submitter": {"lab_uuids": ["72a9c49a-4e17-4df1-96d1-fd19aa46b175"], "fax": "323-442-1224", "last_name": "Farnham", "phone2": "323-442-1145", "google_chat": "Unknown", "phone1": "323-442-8015", "first_name": "Peggy", "@type": ["user", "item"], "labs": ["/labs/72a9c49a-4e17-4df1-96d1-fd19aa46b175"], "lab_name_list": ["Michael Snyder, Stanford"], "skype": "Unknown", "test": "test", "timezone": "UTC-8", "@id": "/users/2eb068c5-b7a6-48ec-aca2-c439e4dabb08/", "email": "pfarnham@usc.edu", "job_title": "PI"}, "product_id": "612044"}, {"lab_uuid": "72a9c49a-4e17-4df1-96d1-fd19aa46b175", "document": {"download": "human_CtBP2_validation_Farnham.pdf", "href": "@@download/document/human_CtBP2_validation_Farnham.pdf", "type": "application/pdf"}, "target_uuid": "0369858d-7f87-42c4-84ff-7a352e825ffa", "date_validated": "Unknown", "antibody_lot_uuid": "b98aaa1c-f594-4059-aefd-621f72a57f9e", "validation_status": "APPROVED", "award_uuid": "df972196-c3c7-4a58-a852-94baa87f9b71", "validation_method": "Knockdowns", "project": "ENCODE2", "caption": "CtBP2 antibody specificity was determined by transfection using CtBP2 siRNA (Dharmacon smart pool) followed by Western blot analysis. siGFP was used for a mock transfection. Western blots were probed with antibodies against CtBP2 (BD 612044), CtBP1 (BD 612042) and actin. As shown, CtBP2 protein levels were reduced when treated with siCtBP2, but not in the mock sample. CtBP1 protein levels as well as the loading control actin remained the same.", "lab": {"award_number_list": ["RC2HG005602", "U54HG004558", "U54HG006996"], "fax": "650-725-1534", "institute_name": "Stanford University", "name": "Michael Snyder, Stanford", "city": "Stanford", "phone2": "650-736-8099", "address1": "Department of Genetics", "address2": "300 Pasteur Drive; MC5120", "phone1": "650-723-4668", "award_uuids": ["95839ec3-9b83-44d9-a98a-026a9fd01bda", "df972196-c3c7-4a58-a852-94baa87f9b71", "4d462953-2da5-4fcf-a695-7206f2d5cf45"], "url": null, "state": "CA", "postal_code": "94305-5120", "institute_label": "Stanford", "test": "test", "awards": ["/awards/95839ec3-9b83-44d9-a98a-026a9fd01bda", "/awards/df972196-c3c7-4a58-a852-94baa87f9b71", "/awards/4d462953-2da5-4fcf-a695-7206f2d5cf45"], "pi_name": null, "country": "USA", "@id": "/labs/72a9c49a-4e17-4df1-96d1-fd19aa46b175/", "@type": ["lab", "item"], "lab_name": null}, "award": {"award_name": null, "description": "Mapping regulatory information throughout the human genome is essential for both determining the function of the human genome and for understanding human development and disease. We propose to establish a high throughput production facility to identify the DNA binding sites for transcription factors and chromatin modifying proteins in human cells. Transcription factor targets, chromatin modifications and modifying enzymes will be mapped across the genome using chromatin immunoprecipitation experiments. This binding information will be validated and analyzed to ensure that it is of high quality and then compared to other genomic annotations. Our three aims are: 1) To establish a production facility for mapping targets of transcription factors. The pipeline will consist of five components: a) cell preparation, b) antibody validation and chromatin immunoprecipitation, c) DNA target analysis, d) target validation and e) sample tracking and databases. 2) To use the pipeline to map targets of 600 transcription factors, chromatin modifications and chromatin modifying enzymes in at least two major cell lines (HeLa S3 and HepG2), and targets of other factors in additional cell lines. 3) To maintain our results in a database and also make them accessible through public databases. We will interrelate our genomic annotations with those determined by other groups (particularly those from the ENCODE consortium), focusing on correlating patterns of transcription-factor binding and chromatin modifications with gene expression. We expect this project to contribute significant information to the identification of regulatory and functional elements in the human genome.", "end_date": "2013-0630", "pi_first_name": "Michael P", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8325805", "award_end_date": null, "award_start_date": null, "number": "U54HG004558", "project": "ENCODE2", "award_number": null, "start_date": "2007-0928", "test": "test", "award_url": null, "pi_last_name": "Snyder", "@id": "/awards/df972196-c3c7-4a58-a852-94baa87f9b71/", "@type": ["award", "item"], "name": "PRODUCTION CENTER FOR GLOBAL MAPPING OF REGULATORY ELEMENTS"}, "target": {"lab_uuid": "fb0af3d0-3a4c-4e96-b67a-f273fe527b04", "date_created": "2012-09-30", "hgnc_dbxref_list": "ribeye", "target_label": "CTBP2", "uniprot_dbxref_list": "P56545", "@id": "/targets/0369858d-7f87-42c4-84ff-7a352e825ffa/", "award_uuid": "1a4d6443-8e29-4b4a-99dd-f93e72d42418", "award": {"award_name": null, "description": "We propose establishing the UCSC ENCODE Data Coordination Center to collect, organize, store, and provide access to data from the ENCODE project and other related projects, based on the well-established UCSC Genome Browser. The UCSC browser is a robust, flexible platform that integrates a variety of features that make accessing genomic data both convenient and information-rich. As such, it has attained a high level of acceptance throughout the biology, bioinformatics, and biomedical research communities. During the pilot phase of the ENCODE project, involving 1% of the genome, UCSC has been the repository of sequence-based data. Through participation in the pilot phase, we have established cost-effective, high-throughput approaches for incorporating and displaying ENCODE data, and we have developed an effective interface to the consortium for uploading data files and methods documentation to our FTP site. Through ongoing expansion and improvement of our ENCODE browser, we have laid the groundwork for data management in the whole-genome phase of the project. UCSC's specific aims for participation as the ENCODE DCC are as follows: 1) coordinate closely with the ENCODE data providers to collect, store, and manage ENCODE sequence-based functional element data and related metadata; 2) provide and maintain a long-term storage system for data from the ENCODE Consortium and related genomic projects; 3) provide the ENCODE and public research communities with freely-available mechanisms to access, search, and analyze the ENCODE data; 4) employ good software engineering practices to ensure that the ENCODE data are of the highest possible quality; and 5) foster collaboration with the biomedical community. Fundamental to understanding human biology and disease is the ability to identify and locate the functional elements in the human genome sequence. Both the effective presentation of experimental data and the ability to analyze sequence data using computational methods will help elucidate genomic function. Because the UCSC Genome Browser serves both of these purposes, it offers the ideal platform for ENCODE data.", "end_date": "2012-1231", "pi_first_name": "W James", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8332919", "award_end_date": null, "award_start_date": null, "number": "U41HG004568", "project": "ENCODE2", "award_number": null, "start_date": "2007-0928", "test": "test", "award_url": null, "pi_last_name": "Kent", "@id": "/awards/1a4d6443-8e29-4b4a-99dd-f93e72d42418/", "@type": ["award", "item"], "name": "THE UCSC ENCODE DATA COORDINATION CENTER"}, "project": "ENCODE2", "target_gene_name": "CTBP2", "geneid_dbxref_list": 1488, "is_current": true, "lab": {"award_number_list": ["U41HG004568"], "fax": "Unknown", "institute_name": "University of California at Santa Cruz", "name": "Jim Kent, UCSC", "city": "Santa Cruz", "phone2": null, "address1": "Baskin School of Engineering", "address2": "1156 High St.; Mail Stop CBSE/ITI", "phone1": "831-459-1401", "award_uuids": ["1a4d6443-8e29-4b4a-99dd-f93e72d42418"], "url": null, "state": "CA", "postal_code": "95064", "institute_label": "UCSC", "test": "test", "awards": ["/awards/1a4d6443-8e29-4b4a-99dd-f93e72d42418"], "pi_name": null, "country": "USA", "@id": "/labs/fb0af3d0-3a4c-4e96-b67a-f273fe527b04/", "@type": ["lab", "item"], "lab_name": null}, "organism_uuid": "7745b647-ff15-4ff3-9ced-b897d4e2983c", "submitter": {"lab_uuids": ["fb0af3d0-3a4c-4e96-b67a-f273fe527b04", "cfb789b8-46f3-4d59-a2b3-adc39e7df93a"], "fax": "Unknown", "last_name": "Kent", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "831-459-1401", "first_name": "Jim", "@type": ["user", "item"], "labs": ["/labs/fb0af3d0-3a4c-4e96-b67a-f273fe527b04", "/labs/cfb789b8-46f3-4d59-a2b3-adc39e7df93a"], "lab_name_list": ["Jim Kent, UCSC", "Michael Cherry, Stanford"], "skype": "Unknown", "test": "test", "timezone": "UTC-8", "@id": "/users/746eef27-d857-4b38-a469-cac93fb02164/", "email": "kent@soe.ucsc.edu", "job_title": "PI"}, "ensembl_dbxref": "ENSG00000175029", "target_class": "Unknown", "submitter_uuid": "746eef27-d857-4b38-a469-cac93fb02164", "organism": {"scientific_name": "Homo sapiens", "taxon_id": 9606, "organism_name": "human", "@id": "/organisms/7745b647-ff15-4ff3-9ced-b897d4e2983c/", "@type": ["organism", "item"]}, "@type": ["target", "item"], "dbxref": {"geneid": ["1488"], "uniprot": ["P56545"], "hgnc": ["ribeye"]}}, "antibody_lot": {"source_uuid": "409ad2e5-845f-4203-9440-0a55298828a9", "lab_uuid": "72a9c49a-4e17-4df1-96d1-fd19aa46b175", "encode2_dbxref_list": "CtBP2", "dbxref": {"encode2": ["CtBP2"]}, "target_label": "CTBP2", "product_id": "612044", "isotype": "Unknown", "clonality": "monoclonal", "award_uuid": "df972196-c3c7-4a58-a852-94baa87f9b71", "purification": "Unknown", "project": "ENCODE2", "source": {"source_name": "BD Biosciences", "url": "http://www.bdbiosciences.com", "actions": [{"profile": "/profiles/source.json", "href": "", "method": "POST", "name": "edit", "title": "Edit"}], "alias": "BD Biosciences", "test": "test", "@id": "/sources/409ad2e5-845f-4203-9440-0a55298828a9/", "@type": ["source", "item"]}, "antigen_description": "Unknown", "antibody_accession": "ENCAB000AFO", "url": "http://www.bdbiosciences.com/ptProduct.jsp?prodId=229326", "antigen_sequence": "Unknown", "host_organism": "mouse", "lot_id": "Unknown", "submitter_uuid": "2eb068c5-b7a6-48ec-aca2-c439e4dabb08", "@id": "/antibody-lots/b98aaa1c-f594-4059-aefd-621f72a57f9e/", "@type": ["antibody_lot", "item"]}, "lot_id": "Unknown", "submitter_uuid": "2eb068c5-b7a6-48ec-aca2-c439e4dabb08", "@id": "/validations/612e938b-ccbb-4e68-894f-8702edc9181f/", "@type": ["antibody_validation", "item"], "submitter": {"lab_uuids": ["72a9c49a-4e17-4df1-96d1-fd19aa46b175"], "fax": "323-442-1224", "last_name": "Farnham", "phone2": "323-442-1145", "google_chat": "Unknown", "phone1": "323-442-8015", "first_name": "Peggy", "@type": ["user", "item"], "labs": ["/labs/72a9c49a-4e17-4df1-96d1-fd19aa46b175"], "lab_name_list": ["Michael Snyder, Stanford"], "skype": "Unknown", "test": "test", "timezone": "UTC-8", "@id": "/users/2eb068c5-b7a6-48ec-aca2-c439e4dabb08/", "email": "pfarnham@usc.edu", "job_title": "PI"}, "product_id": "612044"}], "project": "ENCODE2", "ucsc alias for cricket reference (sanity)": "CtBP2", "antibody_lot": {"source_uuid": "409ad2e5-845f-4203-9440-0a55298828a9", "lab_uuid": "72a9c49a-4e17-4df1-96d1-fd19aa46b175", "encode2_dbxref_list": "CtBP2", "dbxref": {"encode2": ["CtBP2"]}, "target_label": "CTBP2", "product_id": "612044", "isotype": "Unknown", "clonality": "monoclonal", "award_uuid": "df972196-c3c7-4a58-a852-94baa87f9b71", "purification": "Unknown", "project": "ENCODE2", "source": {"source_name": "BD Biosciences", "url": "http://www.bdbiosciences.com", "actions": [{"profile": "/profiles/source.json", "href": "", "method": "POST", "name": "edit", "title": "Edit"}], "alias": "BD Biosciences", "test": "test", "@id": "/sources/409ad2e5-845f-4203-9440-0a55298828a9/", "@type": ["source", "item"]}, "antigen_description": "Unknown", "antibody_accession": "ENCAB000AFO", "url": "http://www.bdbiosciences.com/ptProduct.jsp?prodId=229326", "antigen_sequence": "Unknown", "host_organism": "mouse", "lot_id": "Unknown", "submitter_uuid": "2eb068c5-b7a6-48ec-aca2-c439e4dabb08", "@id": "/antibody-lots/b98aaa1c-f594-4059-aefd-621f72a57f9e/", "@type": ["antibody_lot", "item"]}, "@id": "/antibodies/2dc67301-67ea-4152-b23c-e71bdfa7f8f8/", "@type": ["antibody_approval", "item"], "target": {"lab_uuid": "fb0af3d0-3a4c-4e96-b67a-f273fe527b04", "date_created": "2012-09-30", "hgnc_dbxref_list": "ribeye", "target_label": "CTBP2", "uniprot_dbxref_list": "P56545", "@id": "/targets/0369858d-7f87-42c4-84ff-7a352e825ffa/", "award_uuid": "1a4d6443-8e29-4b4a-99dd-f93e72d42418", "award": {"award_name": null, "description": "We propose establishing the UCSC ENCODE Data Coordination Center to collect, organize, store, and provide access to data from the ENCODE project and other related projects, based on the well-established UCSC Genome Browser. The UCSC browser is a robust, flexible platform that integrates a variety of features that make accessing genomic data both convenient and information-rich. As such, it has attained a high level of acceptance throughout the biology, bioinformatics, and biomedical research communities. During the pilot phase of the ENCODE project, involving 1% of the genome, UCSC has been the repository of sequence-based data. Through participation in the pilot phase, we have established cost-effective, high-throughput approaches for incorporating and displaying ENCODE data, and we have developed an effective interface to the consortium for uploading data files and methods documentation to our FTP site. Through ongoing expansion and improvement of our ENCODE browser, we have laid the groundwork for data management in the whole-genome phase of the project. UCSC's specific aims for participation as the ENCODE DCC are as follows: 1) coordinate closely with the ENCODE data providers to collect, store, and manage ENCODE sequence-based functional element data and related metadata; 2) provide and maintain a long-term storage system for data from the ENCODE Consortium and related genomic projects; 3) provide the ENCODE and public research communities with freely-available mechanisms to access, search, and analyze the ENCODE data; 4) employ good software engineering practices to ensure that the ENCODE data are of the highest possible quality; and 5) foster collaboration with the biomedical community. Fundamental to understanding human biology and disease is the ability to identify and locate the functional elements in the human genome sequence. Both the effective presentation of experimental data and the ability to analyze sequence data using computational methods will help elucidate genomic function. Because the UCSC Genome Browser serves both of these purposes, it offers the ideal platform for ENCODE data.", "end_date": "2012-1231", "pi_first_name": "W James", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8332919", "award_end_date": null, "award_start_date": null, "number": "U41HG004568", "project": "ENCODE2", "award_number": null, "start_date": "2007-0928", "test": "test", "award_url": null, "pi_last_name": "Kent", "@id": "/awards/1a4d6443-8e29-4b4a-99dd-f93e72d42418/", "@type": ["award", "item"], "name": "THE UCSC ENCODE DATA COORDINATION CENTER"}, "project": "ENCODE2", "target_gene_name": "CTBP2", "geneid_dbxref_list": 1488, "is_current": true, "lab": {"award_number_list": ["U41HG004568"], "fax": "Unknown", "institute_name": "University of California at Santa Cruz", "name": "Jim Kent, UCSC", "city": "Santa Cruz", "phone2": null, "address1": "Baskin School of Engineering", "address2": "1156 High St.; Mail Stop CBSE/ITI", "phone1": "831-459-1401", "award_uuids": ["1a4d6443-8e29-4b4a-99dd-f93e72d42418"], "url": null, "state": "CA", "postal_code": "95064", "institute_label": "UCSC", "test": "test", "awards": ["/awards/1a4d6443-8e29-4b4a-99dd-f93e72d42418"], "pi_name": null, "country": "USA", "@id": "/labs/fb0af3d0-3a4c-4e96-b67a-f273fe527b04/", "@type": ["lab", "item"], "lab_name": null}, "organism_uuid": "7745b647-ff15-4ff3-9ced-b897d4e2983c", "submitter": {"lab_uuids": ["fb0af3d0-3a4c-4e96-b67a-f273fe527b04", "cfb789b8-46f3-4d59-a2b3-adc39e7df93a"], "fax": "Unknown", "last_name": "Kent", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "831-459-1401", "first_name": "Jim", "@type": ["user", "item"], "labs": ["/labs/fb0af3d0-3a4c-4e96-b67a-f273fe527b04", "/labs/cfb789b8-46f3-4d59-a2b3-adc39e7df93a"], "lab_name_list": ["Jim Kent, UCSC", "Michael Cherry, Stanford"], "skype": "Unknown", "test": "test", "timezone": "UTC-8", "@id": "/users/746eef27-d857-4b38-a469-cac93fb02164/", "email": "kent@soe.ucsc.edu", "job_title": "PI"}, "ensembl_dbxref": "ENSG00000175029", "target_class": "Unknown", "submitter_uuid": "746eef27-d857-4b38-a469-cac93fb02164", "organism": {"scientific_name": "Homo sapiens", "taxon_id": 9606, "organism_name": "human", "@id": "/organisms/7745b647-ff15-4ff3-9ced-b897d4e2983c/", "@type": ["organism", "item"]}, "@type": ["target", "item"], "dbxref": {"geneid": ["1488"], "uniprot": ["P56545"], "hgnc": ["ribeye"]}}}], "experiments": [{"files": [], "lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "is_obsolete": false, "date_submitted": "2013-05-02", "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "replicates": [{"platform_uuid": "03a96eaf-75da-44e2-b4f6-e349e8c3655e", "biological_replicate_number": 2, "biosample_accession (sanity)": "ENCBS086RNA", "biosample (sanity)": "139T", "platform": {"encode2_dbxref_list": "Illumina_HiSeq_2000", "gpl_ids": ["GPL11154", "GPL13112"], "description": "Illumina HiSeq 2000", "url": "http://www.illumina.com/documents/products/datasheets/datasheet_hiseq2000.pdf", "obi_dbxref_list": "new term will be requested", "test": "test", "dbxref": {"obi": ["new term will be requested"], "geo": ["GPL11154", "GPL13112"], "encode2": ["Illumina_HiSeq_2000"]}, "@id": "/platforms/03a96eaf-75da-44e2-b4f6-e349e8c3655e/", "@type": ["platform", "item"]}, "assay_type (obi)": "OBI_0001864", "@id": "/replicates/c8606394-9b18-4fbf-9a9c-d2b109d71c01/", "technical_replicate_number": 1, "library": {"document_uuids": [], "lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "nucleic_acid_type": "RNA", "strand_specificity": true, "@id": "/libraries/4f1b2c43-3769-4fb4-ebf0-911b0ae7fdbe/", "size_range": "&gt; 200", "date_submitted": "2013-05-02", "accession": "ENCLB294ZZZ", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "nucleic_acid_type_so": "SO:0000356", "documents": [], "biosample": {"accession": "ENCBS086RNA", "product_url": "http://www.biochain.com/index.php/catalogsearch/advanced/result/?catalog_number=R1244108-10", "donor_uuid": "f6277156-eb8e-473d-95eb-761ab6860793", "biosample_description": "Human fetal whole eye", "starting_amount": "Unknown", "lot_id": "B608426", "culture_harvest_date": "N/A", "related_biosample_uuid": "", "documents": [{"lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "document_type": "data sheet", "document": {"download": "B608426.pdf", "href": "@@download/document/B608426.pdf", "type": "application/pdf"}, "description": "BioChain Human fetal whole eye B608426 data sheet", "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "document_label": "BC_B608426", "test": "test", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "@id": "/documents/e40f45be-9294-42b1-8e2b-66783ea21426/", "@type": ["biosample_document", "item"], "document_name": "Eye_BC_B608426_data_sheet"}], "biosample_type": "Tissue", "passage_number": "N/A", "note": "Purchased as total RNA", "source": {"source_name": "BioChain Institute Inc", "url": "http://www.biochain.com", "actions": [{"profile": "/profiles/source.json", "href": "", "method": "POST", "name": "edit", "title": "Edit"}], "alias": "BioChain", "test": "test", "@id": "/sources/8599bb19-6c08-4610-9ebf-125c5346461e/", "@type": ["source", "item"]}, "dr. thomas gingeras at cold spring harbor laboratory_dbxref_list": "139T", "test": "test", "donor": {"strain background": "N/A", "organism": {"scientific_name": "Homo sapiens", "taxon_id": 9606, "organism_name": "human", "@id": "/organisms/7745b647-ff15-4ff3-9ced-b897d4e2983c/", "@type": ["organism", "item"]}, "age": "20 week fetal", "alias_source_uuids": ["8599bb19-6c08-4610-9ebf-125c5346461e"], "alias_list": ["F32"], "sex": "Female", "health_status": "Normal", "donor_id": "ENCDO064AAA", "organism_uuid": "7745b647-ff15-4ff3-9ced-b897d4e2983c", "test": "test", "@id": "/donors/f6277156-eb8e-473d-95eb-761ab6860793/", "@type": ["donor", "item"], "ethnicity": "Unknown"}, "lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "treatment_uuids": [], "constructs": [], "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "treatments": [], "dbxref": {"dr. richard myers at hudson alpha": [], "dr. thomas gingeras at cold spring harbor laboratory": ["139T"], "encode2": []}, "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "date_obtained": "Unknown", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "@id": "/biosamples/af037bca-053a-4c09-8a47-be232222df45/", "document_uuids": ["e40f45be-9294-42b1-8e2b-66783ea21426"], "culture_start_date": "N/A", "product_id": "R1244108-10-D02", "biosample_term_id": "UBERON:0000019", "construct_uuids": [], "biosample_term_name": "camera-type eye", "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "source_uuid": "8599bb19-6c08-4610-9ebf-125c5346461e", "@type": ["biosample", "item"], "related_biosample_accession": ""}, "dr. thomas gingeras at cold spring harbor laboratory_dbxref_list": "SID38460", "library_description": "139T; CSHL RAMPAGE Human fetal whole eye", "biosample_uuid": "af037bca-053a-4c09-8a47-be232222df45", "@type": ["library", "item"], "dbxref": {"dr. barbara wold at caltech": [], "dr. richard myers at hudson alpha": [], "dr. thomas gingeras at cold spring harbor laboratory": ["SID38460"]}}, "assay_uuid": "13c3cd26-a8b0-479a-a8c5-736b7575731f", "library_uuid": "4f1b2c43-3769-4fb4-ebf0-911b0ae7fdbe", "library id (sanity)": "ENCLB294ZZZ", "@type": ["replicate", "item"], "assay": {"dbxref": {"obi": ["http://purl.obolibrary.org/obo/OBI_0001864"], "encode2": []}, "obi_dbxref_list": "http://purl.obolibrary.org/obo/OBI_0001864", "assay_name": "RAMPAGE", "test": "test", "assay_description": "RNA Annotation and Mapping of Promoters for the Analysis of Gene Expression assay", "@id": "/assays/13c3cd26-a8b0-479a-a8c5-736b7575731f/", "@type": ["assay", "item"]}, "experiment_accession": "ENCSR000AHC", "is_complete": false}, {"platform_uuid": "03a96eaf-75da-44e2-b4f6-e349e8c3655e", "biological_replicate_number": 1, "biosample_accession (sanity)": "ENCBS085RNA", "biosample (sanity)": "138T", "platform": {"encode2_dbxref_list": "Illumina_HiSeq_2000", "gpl_ids": ["GPL11154", "GPL13112"], "description": "Illumina HiSeq 2000", "url": "http://www.illumina.com/documents/products/datasheets/datasheet_hiseq2000.pdf", "obi_dbxref_list": "new term will be requested", "test": "test", "dbxref": {"obi": ["new term will be requested"], "geo": ["GPL11154", "GPL13112"], "encode2": ["Illumina_HiSeq_2000"]}, "@id": "/platforms/03a96eaf-75da-44e2-b4f6-e349e8c3655e/", "@type": ["platform", "item"]}, "assay_type (obi)": "OBI_0001864", "@id": "/replicates/4551f103-87e2-42da-8daa-d8da1d97dd5a/", "technical_replicate_number": 1, "library": {"document_uuids": [], "lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "nucleic_acid_type": "RNA", "strand_specificity": true, "@id": "/libraries/9d7823ef-eb46-4838-a393-0bd11f4445dc/", "size_range": "&gt; 200", "date_submitted": "2013-05-02", "accession": "ENCLB293ZZZ", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "nucleic_acid_type_so": "SO:0000356", "documents": [], "biosample": {"accession": "ENCBS085RNA", "product_url": "http://www.biochain.com/index.php/catalogsearch/advanced/result/?catalog_number=R1244108-10", "donor_uuid": "998ef4ba-5aca-4dd0-9ef2-fa030628a115", "biosample_description": "Human fetal whole eye", "starting_amount": "Unknown", "lot_id": "B608427", "culture_harvest_date": "N/A", "related_biosample_uuid": "", "documents": [{"lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "document_type": "data sheet", "document": {"download": "B608427.pdf", "href": "@@download/document/B608427.pdf", "type": "application/pdf"}, "description": "BioChain Human fetal whole eye B608427 data sheet", "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "document_label": "BC_B608427", "test": "test", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "@id": "/documents/243a7e03-3b82-478f-aeba-639e9453910d/", "@type": ["biosample_document", "item"], "document_name": "Eye_BC_B608427_data_sheet"}], "biosample_type": "Tissue", "passage_number": "N/A", "note": "Purchased as total RNA", "source": {"source_name": "BioChain Institute Inc", "url": "http://www.biochain.com", "actions": [{"profile": "/profiles/source.json", "href": "", "method": "POST", "name": "edit", "title": "Edit"}], "alias": "BioChain", "test": "test", "@id": "/sources/8599bb19-6c08-4610-9ebf-125c5346461e/", "@type": ["source", "item"]}, "dr. thomas gingeras at cold spring harbor laboratory_dbxref_list": "138T", "test": "test", "donor": {"strain background": "N/A", "organism": {"scientific_name": "Homo sapiens", "taxon_id": 9606, "organism_name": "human", "@id": "/organisms/7745b647-ff15-4ff3-9ced-b897d4e2983c/", "@type": ["organism", "item"]}, "age": "24 week fetal", "alias_source_uuids": ["8599bb19-6c08-4610-9ebf-125c5346461e"], "alias_list": ["F63"], "sex": "Female", "health_status": "Normal", "donor_id": "ENCDO067AAA", "organism_uuid": "7745b647-ff15-4ff3-9ced-b897d4e2983c", "test": "test", "@id": "/donors/998ef4ba-5aca-4dd0-9ef2-fa030628a115/", "@type": ["donor", "item"], "ethnicity": "Unknown"}, "lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "treatment_uuids": [], "constructs": [], "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "treatments": [], "dbxref": {"dr. richard myers at hudson alpha": [], "dr. thomas gingeras at cold spring harbor laboratory": ["138T"], "encode2": []}, "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "date_obtained": "Unknown", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "@id": "/biosamples/12cc1219-9c09-400c-9a9a-8d90953cf2b7/", "document_uuids": ["243a7e03-3b82-478f-aeba-639e9453910d"], "culture_start_date": "N/A", "product_id": "R1244108-10-D02", "biosample_term_id": "UBERON:0000019", "construct_uuids": [], "biosample_term_name": "camera-type eye", "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "source_uuid": "8599bb19-6c08-4610-9ebf-125c5346461e", "@type": ["biosample", "item"], "related_biosample_accession": ""}, "dr. thomas gingeras at cold spring harbor laboratory_dbxref_list": "SID38459", "library_description": "138T; CSHL RAMPAGE Human fetal whole eye", "biosample_uuid": "12cc1219-9c09-400c-9a9a-8d90953cf2b7", "@type": ["library", "item"], "dbxref": {"dr. barbara wold at caltech": [], "dr. richard myers at hudson alpha": [], "dr. thomas gingeras at cold spring harbor laboratory": ["SID38459"]}}, "assay_uuid": "13c3cd26-a8b0-479a-a8c5-736b7575731f", "library_uuid": "9d7823ef-eb46-4838-a393-0bd11f4445dc", "library id (sanity)": "ENCLB293ZZZ", "@type": ["replicate", "item"], "assay": {"dbxref": {"obi": ["http://purl.obolibrary.org/obo/OBI_0001864"], "encode2": []}, "obi_dbxref_list": "http://purl.obolibrary.org/obo/OBI_0001864", "assay_name": "RAMPAGE", "test": "test", "assay_description": "RNA Annotation and Mapping of Promoters for the Analysis of Gene Expression assay", "@id": "/assays/13c3cd26-a8b0-479a-a8c5-736b7575731f/", "@type": ["assay", "item"]}, "experiment_accession": "ENCSR000AHC", "is_complete": false}], "dataset_description": "CSHL RAMPAGE RNA-seq Human fetal whole eye 138T ; 139T", "file_uuids": [], "dataset_type": "experiment", "experiment_control_uuids": [], "replicate_uuids": ["c8606394-9b18-4fbf-9a9c-d2b109d71c01", "4551f103-87e2-42da-8daa-d8da1d97dd5a"], "dbxref": {"geo": [], "encode2": []}, "@id": "/experiments/7779bea9-6a73-47da-b508-35ef1b28eaa4/", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "dataset_accession": "ENCSR000AHC", "document_list": [], "controls": [], "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "@type": ["experiment", "item"]}, {"files": [], "lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "is_obsolete": false, "date_submitted": "2013-05-02", "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "replicates": [{"platform_uuid": "03a96eaf-75da-44e2-b4f6-e349e8c3655e", "biological_replicate_number": 2, "biosample_accession (sanity)": "ENCBS086RNA", "biosample (sanity)": "139T", "platform": {"encode2_dbxref_list": "Illumina_HiSeq_2000", "gpl_ids": ["GPL11154", "GPL13112"], "description": "Illumina HiSeq 2000", "url": "http://www.illumina.com/documents/products/datasheets/datasheet_hiseq2000.pdf", "obi_dbxref_list": "new term will be requested", "test": "test", "dbxref": {"obi": ["new term will be requested"], "geo": ["GPL11154", "GPL13112"], "encode2": ["Illumina_HiSeq_2000"]}, "@id": "/platforms/03a96eaf-75da-44e2-b4f6-e349e8c3655e/", "@type": ["platform", "item"]}, "assay_type (obi)": "OBI_0001271", "@id": "/replicates/700e8657-e2de-4da0-aa19-408028185dc0/", "technical_replicate_number": 1, "library": {"document_uuids": [], "lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "documents": [], "strand_specificity": true, "@type": ["library", "item"], "biosample_uuid": "af037bca-053a-4c09-8a47-be232222df45", "nucleic_acid_type_so": "SO:0000356", "date_submitted": "2013-05-02", "accession": "ENCLB214ZZZ", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "nucleic_acid_type": "RNA", "biosample": {"accession": "ENCBS086RNA", "product_url": "http://www.biochain.com/index.php/catalogsearch/advanced/result/?catalog_number=R1244108-10", "donor_uuid": "f6277156-eb8e-473d-95eb-761ab6860793", "biosample_description": "Human fetal whole eye", "starting_amount": "Unknown", "lot_id": "B608426", "culture_harvest_date": "N/A", "related_biosample_uuid": "", "documents": [{"lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "document_type": "data sheet", "document": {"download": "B608426.pdf", "href": "@@download/document/B608426.pdf", "type": "application/pdf"}, "description": "BioChain Human fetal whole eye B608426 data sheet", "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "document_label": "BC_B608426", "test": "test", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "@id": "/documents/e40f45be-9294-42b1-8e2b-66783ea21426/", "@type": ["biosample_document", "item"], "document_name": "Eye_BC_B608426_data_sheet"}], "biosample_type": "Tissue", "passage_number": "N/A", "note": "Purchased as total RNA", "source": {"source_name": "BioChain Institute Inc", "url": "http://www.biochain.com", "actions": [{"profile": "/profiles/source.json", "href": "", "method": "POST", "name": "edit", "title": "Edit"}], "alias": "BioChain", "test": "test", "@id": "/sources/8599bb19-6c08-4610-9ebf-125c5346461e/", "@type": ["source", "item"]}, "dr. thomas gingeras at cold spring harbor laboratory_dbxref_list": "139T", "test": "test", "donor": {"strain background": "N/A", "organism": {"scientific_name": "Homo sapiens", "taxon_id": 9606, "organism_name": "human", "@id": "/organisms/7745b647-ff15-4ff3-9ced-b897d4e2983c/", "@type": ["organism", "item"]}, "age": "20 week fetal", "alias_source_uuids": ["8599bb19-6c08-4610-9ebf-125c5346461e"], "alias_list": ["F32"], "sex": "Female", "health_status": "Normal", "donor_id": "ENCDO064AAA", "organism_uuid": "7745b647-ff15-4ff3-9ced-b897d4e2983c", "test": "test", "@id": "/donors/f6277156-eb8e-473d-95eb-761ab6860793/", "@type": ["donor", "item"], "ethnicity": "Unknown"}, "lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "treatment_uuids": [], "constructs": [], "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "treatments": [], "dbxref": {"dr. richard myers at hudson alpha": [], "dr. thomas gingeras at cold spring harbor laboratory": ["139T"], "encode2": []}, "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "date_obtained": "Unknown", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "@id": "/biosamples/af037bca-053a-4c09-8a47-be232222df45/", "document_uuids": ["e40f45be-9294-42b1-8e2b-66783ea21426"], "culture_start_date": "N/A", "product_id": "R1244108-10-D02", "biosample_term_id": "UBERON:0000019", "construct_uuids": [], "biosample_term_name": "camera-type eye", "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "source_uuid": "8599bb19-6c08-4610-9ebf-125c5346461e", "@type": ["biosample", "item"], "related_biosample_accession": ""}, "dr. thomas gingeras at cold spring harbor laboratory_dbxref_list": "SID38173", "size_range": "&gt;200", "library_description": "139T; CSHL Long RNA-seq Human fetal whole eye", "@id": "/libraries/9cd741e9-56c9-42d8-9609-604c026ee7ff/", "library_size_selection_method": "gel", "dbxref": {"dr. barbara wold at caltech": [], "dr. richard myers at hudson alpha": [], "dr. thomas gingeras at cold spring harbor laboratory": ["SID38173"]}}, "assay_uuid": "0963a733-b6ee-453f-ba2c-73288efe19a7", "library_uuid": "9cd741e9-56c9-42d8-9609-604c026ee7ff", "library id (sanity)": "ENCLB214ZZZ", "@type": ["replicate", "item"], "assay": {"encode2_dbxref_list": "RnaSeq", "assay_description": "an assay that uses high-throughput sequencing technologies to sequence cDNA in order to get information about a sample's RNA content.", "obi_dbxref_list": "http://purl.obolibrary.org/obo/OBI_0001271", "assay_name": "RNA-seq", "test": "test", "dbxref": {"obi": ["http://purl.obolibrary.org/obo/OBI_0001271"], "encode2": ["RnaSeq"]}, "@id": "/assays/0963a733-b6ee-453f-ba2c-73288efe19a7/", "@type": ["assay", "item"]}, "experiment_accession": "ENCSR000AFO", "is_complete": false}, {"platform_uuid": "03a96eaf-75da-44e2-b4f6-e349e8c3655e", "biological_replicate_number": 1, "biosample_accession (sanity)": "ENCBS085RNA", "biosample (sanity)": "138T", "platform": {"encode2_dbxref_list": "Illumina_HiSeq_2000", "gpl_ids": ["GPL11154", "GPL13112"], "description": "Illumina HiSeq 2000", "url": "http://www.illumina.com/documents/products/datasheets/datasheet_hiseq2000.pdf", "obi_dbxref_list": "new term will be requested", "test": "test", "dbxref": {"obi": ["new term will be requested"], "geo": ["GPL11154", "GPL13112"], "encode2": ["Illumina_HiSeq_2000"]}, "@id": "/platforms/03a96eaf-75da-44e2-b4f6-e349e8c3655e/", "@type": ["platform", "item"]}, "assay_type (obi)": "OBI_0001271", "@id": "/replicates/95202968-24f5-47d7-b4fb-44ac41a3bb5f/", "technical_replicate_number": 1, "library": {"document_uuids": [], "lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "documents": [], "strand_specificity": true, "@type": ["library", "item"], "biosample_uuid": "12cc1219-9c09-400c-9a9a-8d90953cf2b7", "nucleic_acid_type_so": "SO:0000356", "date_submitted": "2013-05-02", "accession": "ENCLB213ZZZ", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "nucleic_acid_type": "RNA", "biosample": {"accession": "ENCBS085RNA", "product_url": "http://www.biochain.com/index.php/catalogsearch/advanced/result/?catalog_number=R1244108-10", "donor_uuid": "998ef4ba-5aca-4dd0-9ef2-fa030628a115", "biosample_description": "Human fetal whole eye", "starting_amount": "Unknown", "lot_id": "B608427", "culture_harvest_date": "N/A", "related_biosample_uuid": "", "documents": [{"lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "document_type": "data sheet", "document": {"download": "B608427.pdf", "href": "@@download/document/B608427.pdf", "type": "application/pdf"}, "description": "BioChain Human fetal whole eye B608427 data sheet", "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "document_label": "BC_B608427", "test": "test", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "@id": "/documents/243a7e03-3b82-478f-aeba-639e9453910d/", "@type": ["biosample_document", "item"], "document_name": "Eye_BC_B608427_data_sheet"}], "biosample_type": "Tissue", "passage_number": "N/A", "note": "Purchased as total RNA", "source": {"source_name": "BioChain Institute Inc", "url": "http://www.biochain.com", "actions": [{"profile": "/profiles/source.json", "href": "", "method": "POST", "name": "edit", "title": "Edit"}], "alias": "BioChain", "test": "test", "@id": "/sources/8599bb19-6c08-4610-9ebf-125c5346461e/", "@type": ["source", "item"]}, "dr. thomas gingeras at cold spring harbor laboratory_dbxref_list": "138T", "test": "test", "donor": {"strain background": "N/A", "organism": {"scientific_name": "Homo sapiens", "taxon_id": 9606, "organism_name": "human", "@id": "/organisms/7745b647-ff15-4ff3-9ced-b897d4e2983c/", "@type": ["organism", "item"]}, "age": "24 week fetal", "alias_source_uuids": ["8599bb19-6c08-4610-9ebf-125c5346461e"], "alias_list": ["F63"], "sex": "Female", "health_status": "Normal", "donor_id": "ENCDO067AAA", "organism_uuid": "7745b647-ff15-4ff3-9ced-b897d4e2983c", "test": "test", "@id": "/donors/998ef4ba-5aca-4dd0-9ef2-fa030628a115/", "@type": ["donor", "item"], "ethnicity": "Unknown"}, "lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "treatment_uuids": [], "constructs": [], "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "treatments": [], "dbxref": {"dr. richard myers at hudson alpha": [], "dr. thomas gingeras at cold spring harbor laboratory": ["138T"], "encode2": []}, "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "date_obtained": "Unknown", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "@id": "/biosamples/12cc1219-9c09-400c-9a9a-8d90953cf2b7/", "document_uuids": ["243a7e03-3b82-478f-aeba-639e9453910d"], "culture_start_date": "N/A", "product_id": "R1244108-10-D02", "biosample_term_id": "UBERON:0000019", "construct_uuids": [], "biosample_term_name": "camera-type eye", "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "source_uuid": "8599bb19-6c08-4610-9ebf-125c5346461e", "@type": ["biosample", "item"], "related_biosample_accession": ""}, "dr. thomas gingeras at cold spring harbor laboratory_dbxref_list": "SID38172", "size_range": "&gt;200", "library_description": "138T; CSHL Long RNA-seq Human fetal whole eye", "@id": "/libraries/86030ca4-0244-481b-fd49-d536d5ee60da/", "library_size_selection_method": "gel", "dbxref": {"dr. barbara wold at caltech": [], "dr. richard myers at hudson alpha": [], "dr. thomas gingeras at cold spring harbor laboratory": ["SID38172"]}}, "assay_uuid": "0963a733-b6ee-453f-ba2c-73288efe19a7", "library_uuid": "86030ca4-0244-481b-fd49-d536d5ee60da", "library id (sanity)": "ENCLB213ZZZ", "@type": ["replicate", "item"], "assay": {"encode2_dbxref_list": "RnaSeq", "assay_description": "an assay that uses high-throughput sequencing technologies to sequence cDNA in order to get information about a sample's RNA content.", "obi_dbxref_list": "http://purl.obolibrary.org/obo/OBI_0001271", "assay_name": "RNA-seq", "test": "test", "dbxref": {"obi": ["http://purl.obolibrary.org/obo/OBI_0001271"], "encode2": ["RnaSeq"]}, "@id": "/assays/0963a733-b6ee-453f-ba2c-73288efe19a7/", "@type": ["assay", "item"]}, "experiment_accession": "ENCSR000AFO", "is_complete": false}], "dataset_description": "CSHL Long RNA-seq Human fetal whole eye 138T ; 139T", "file_uuids": [], "dataset_type": "experiment", "experiment_control_uuids": [], "replicate_uuids": ["700e8657-e2de-4da0-aa19-408028185dc0", "95202968-24f5-47d7-b4fb-44ac41a3bb5f"], "dbxref": {"geo": [], "encode2": []}, "@id": "/experiments/39c6c55e-f1cf-449e-cfab-07b11fc5ae7c/", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "dataset_accession": "ENCSR000AFO", "document_list": [], "controls": [], "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "@type": ["experiment", "item"]}, {"files": [], "lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "is_obsolete": false, "date_submitted": "2013-05-02", "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "replicates": [{"platform_uuid": "03a96eaf-75da-44e2-b4f6-e349e8c3655e", "biological_replicate_number": 1, "biosample_accession (sanity)": "ENCBS085RNA", "biosample (sanity)": "138T", "platform": {"encode2_dbxref_list": "Illumina_HiSeq_2000", "gpl_ids": ["GPL11154", "GPL13112"], "description": "Illumina HiSeq 2000", "url": "http://www.illumina.com/documents/products/datasheets/datasheet_hiseq2000.pdf", "obi_dbxref_list": "new term will be requested", "test": "test", "dbxref": {"obi": ["new term will be requested"], "geo": ["GPL11154", "GPL13112"], "encode2": ["Illumina_HiSeq_2000"]}, "@id": "/platforms/03a96eaf-75da-44e2-b4f6-e349e8c3655e/", "@type": ["platform", "item"]}, "assay_type (obi)": "OBI_0001271", "@id": "/replicates/46b9107d-3530-490e-c712-e913943ceadf/", "technical_replicate_number": 1, "library": {"document_uuids": [], "lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "nucleic_acid_type": "Small", "strand_specificity": true, "@id": "/libraries/8d1d0aa9-45e4-4b29-d629-ec6feea98a4a/", "size_range": "&lt; 200", "date_submitted": "2013-05-02", "accession": "ENCLB253ZZZ", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "nucleic_acid_type_so": "SO:0000356", "documents": [], "biosample": {"accession": "ENCBS085RNA", "product_url": "http://www.biochain.com/index.php/catalogsearch/advanced/result/?catalog_number=R1244108-10", "donor_uuid": "998ef4ba-5aca-4dd0-9ef2-fa030628a115", "biosample_description": "Human fetal whole eye", "starting_amount": "Unknown", "lot_id": "B608427", "culture_harvest_date": "N/A", "related_biosample_uuid": "", "documents": [{"lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "document_type": "data sheet", "document": {"download": "B608427.pdf", "href": "@@download/document/B608427.pdf", "type": "application/pdf"}, "description": "BioChain Human fetal whole eye B608427 data sheet", "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "document_label": "BC_B608427", "test": "test", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "@id": "/documents/243a7e03-3b82-478f-aeba-639e9453910d/", "@type": ["biosample_document", "item"], "document_name": "Eye_BC_B608427_data_sheet"}], "biosample_type": "Tissue", "passage_number": "N/A", "note": "Purchased as total RNA", "source": {"source_name": "BioChain Institute Inc", "url": "http://www.biochain.com", "actions": [{"profile": "/profiles/source.json", "href": "", "method": "POST", "name": "edit", "title": "Edit"}], "alias": "BioChain", "test": "test", "@id": "/sources/8599bb19-6c08-4610-9ebf-125c5346461e/", "@type": ["source", "item"]}, "dr. thomas gingeras at cold spring harbor laboratory_dbxref_list": "138T", "test": "test", "donor": {"strain background": "N/A", "organism": {"scientific_name": "Homo sapiens", "taxon_id": 9606, "organism_name": "human", "@id": "/organisms/7745b647-ff15-4ff3-9ced-b897d4e2983c/", "@type": ["organism", "item"]}, "age": "24 week fetal", "alias_source_uuids": ["8599bb19-6c08-4610-9ebf-125c5346461e"], "alias_list": ["F63"], "sex": "Female", "health_status": "Normal", "donor_id": "ENCDO067AAA", "organism_uuid": "7745b647-ff15-4ff3-9ced-b897d4e2983c", "test": "test", "@id": "/donors/998ef4ba-5aca-4dd0-9ef2-fa030628a115/", "@type": ["donor", "item"], "ethnicity": "Unknown"}, "lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "treatment_uuids": [], "constructs": [], "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "treatments": [], "dbxref": {"dr. richard myers at hudson alpha": [], "dr. thomas gingeras at cold spring harbor laboratory": ["138T"], "encode2": []}, "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "date_obtained": "Unknown", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "@id": "/biosamples/12cc1219-9c09-400c-9a9a-8d90953cf2b7/", "document_uuids": ["243a7e03-3b82-478f-aeba-639e9453910d"], "culture_start_date": "N/A", "product_id": "R1244108-10-D02", "biosample_term_id": "UBERON:0000019", "construct_uuids": [], "biosample_term_name": "camera-type eye", "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "source_uuid": "8599bb19-6c08-4610-9ebf-125c5346461e", "@type": ["biosample", "item"], "related_biosample_accession": ""}, "dr. thomas gingeras at cold spring harbor laboratory_dbxref_list": "FAIL", "library_description": "138T; CSHL Small RNA-seq Human fetal whole eye", "biosample_uuid": "12cc1219-9c09-400c-9a9a-8d90953cf2b7", "@type": ["library", "item"], "dbxref": {"dr. barbara wold at caltech": [], "dr. richard myers at hudson alpha": [], "dr. thomas gingeras at cold spring harbor laboratory": ["FAIL"]}}, "assay_uuid": "0963a733-b6ee-453f-ba2c-73288efe19a7", "library_uuid": "8d1d0aa9-45e4-4b29-d629-ec6feea98a4a", "library id (sanity)": "ENCLB253ZZZ", "@type": ["replicate", "item"], "assay": {"encode2_dbxref_list": "RnaSeq", "assay_description": "an assay that uses high-throughput sequencing technologies to sequence cDNA in order to get information about a sample's RNA content.", "obi_dbxref_list": "http://purl.obolibrary.org/obo/OBI_0001271", "assay_name": "RNA-seq", "test": "test", "dbxref": {"obi": ["http://purl.obolibrary.org/obo/OBI_0001271"], "encode2": ["RnaSeq"]}, "@id": "/assays/0963a733-b6ee-453f-ba2c-73288efe19a7/", "@type": ["assay", "item"]}, "experiment_accession": "ENCSR000AGI", "is_complete": false}, {"platform_uuid": "03a96eaf-75da-44e2-b4f6-e349e8c3655e", "biological_replicate_number": 2, "biosample_accession (sanity)": "ENCBS086RNA", "biosample (sanity)": "139T", "platform": {"encode2_dbxref_list": "Illumina_HiSeq_2000", "gpl_ids": ["GPL11154", "GPL13112"], "description": "Illumina HiSeq 2000", "url": "http://www.illumina.com/documents/products/datasheets/datasheet_hiseq2000.pdf", "obi_dbxref_list": "new term will be requested", "test": "test", "dbxref": {"obi": ["new term will be requested"], "geo": ["GPL11154", "GPL13112"], "encode2": ["Illumina_HiSeq_2000"]}, "@id": "/platforms/03a96eaf-75da-44e2-b4f6-e349e8c3655e/", "@type": ["platform", "item"]}, "assay_type (obi)": "OBI_0001271", "@id": "/replicates/f7d9d2dd-bd3b-44ec-ac96-e1bfc8b90bc6/", "technical_replicate_number": 1, "library": {"document_uuids": [], "lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "nucleic_acid_type": "Small", "strand_specificity": true, "@id": "/libraries/e51be54d-6bad-4ba7-b8e3-aaaf733205a7/", "size_range": "&lt; 200", "date_submitted": "2013-05-02", "accession": "ENCLB254ZZZ", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "nucleic_acid_type_so": "SO:0000356", "documents": [], "biosample": {"accession": "ENCBS086RNA", "product_url": "http://www.biochain.com/index.php/catalogsearch/advanced/result/?catalog_number=R1244108-10", "donor_uuid": "f6277156-eb8e-473d-95eb-761ab6860793", "biosample_description": "Human fetal whole eye", "starting_amount": "Unknown", "lot_id": "B608426", "culture_harvest_date": "N/A", "related_biosample_uuid": "", "documents": [{"lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "document_type": "data sheet", "document": {"download": "B608426.pdf", "href": "@@download/document/B608426.pdf", "type": "application/pdf"}, "description": "BioChain Human fetal whole eye B608426 data sheet", "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "document_label": "BC_B608426", "test": "test", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "@id": "/documents/e40f45be-9294-42b1-8e2b-66783ea21426/", "@type": ["biosample_document", "item"], "document_name": "Eye_BC_B608426_data_sheet"}], "biosample_type": "Tissue", "passage_number": "N/A", "note": "Purchased as total RNA", "source": {"source_name": "BioChain Institute Inc", "url": "http://www.biochain.com", "actions": [{"profile": "/profiles/source.json", "href": "", "method": "POST", "name": "edit", "title": "Edit"}], "alias": "BioChain", "test": "test", "@id": "/sources/8599bb19-6c08-4610-9ebf-125c5346461e/", "@type": ["source", "item"]}, "dr. thomas gingeras at cold spring harbor laboratory_dbxref_list": "139T", "test": "test", "donor": {"strain background": "N/A", "organism": {"scientific_name": "Homo sapiens", "taxon_id": 9606, "organism_name": "human", "@id": "/organisms/7745b647-ff15-4ff3-9ced-b897d4e2983c/", "@type": ["organism", "item"]}, "age": "20 week fetal", "alias_source_uuids": ["8599bb19-6c08-4610-9ebf-125c5346461e"], "alias_list": ["F32"], "sex": "Female", "health_status": "Normal", "donor_id": "ENCDO064AAA", "organism_uuid": "7745b647-ff15-4ff3-9ced-b897d4e2983c", "test": "test", "@id": "/donors/f6277156-eb8e-473d-95eb-761ab6860793/", "@type": ["donor", "item"], "ethnicity": "Unknown"}, "lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "treatment_uuids": [], "constructs": [], "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "treatments": [], "dbxref": {"dr. richard myers at hudson alpha": [], "dr. thomas gingeras at cold spring harbor laboratory": ["139T"], "encode2": []}, "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "date_obtained": "Unknown", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "@id": "/biosamples/af037bca-053a-4c09-8a47-be232222df45/", "document_uuids": ["e40f45be-9294-42b1-8e2b-66783ea21426"], "culture_start_date": "N/A", "product_id": "R1244108-10-D02", "biosample_term_id": "UBERON:0000019", "construct_uuids": [], "biosample_term_name": "camera-type eye", "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "source_uuid": "8599bb19-6c08-4610-9ebf-125c5346461e", "@type": ["biosample", "item"], "related_biosample_accession": ""}, "dr. thomas gingeras at cold spring harbor laboratory_dbxref_list": "SID38495", "library_description": "139T; CSHL Small RNA-seq Human fetal whole eye", "biosample_uuid": "af037bca-053a-4c09-8a47-be232222df45", "@type": ["library", "item"], "dbxref": {"dr. barbara wold at caltech": [], "dr. richard myers at hudson alpha": [], "dr. thomas gingeras at cold spring harbor laboratory": ["SID38495"]}}, "assay_uuid": "0963a733-b6ee-453f-ba2c-73288efe19a7", "library_uuid": "e51be54d-6bad-4ba7-b8e3-aaaf733205a7", "library id (sanity)": "ENCLB254ZZZ", "@type": ["replicate", "item"], "assay": {"encode2_dbxref_list": "RnaSeq", "assay_description": "an assay that uses high-throughput sequencing technologies to sequence cDNA in order to get information about a sample's RNA content.", "obi_dbxref_list": "http://purl.obolibrary.org/obo/OBI_0001271", "assay_name": "RNA-seq", "test": "test", "dbxref": {"obi": ["http://purl.obolibrary.org/obo/OBI_0001271"], "encode2": ["RnaSeq"]}, "@id": "/assays/0963a733-b6ee-453f-ba2c-73288efe19a7/", "@type": ["assay", "item"]}, "experiment_accession": "ENCSR000AGI", "is_complete": false}], "dataset_description": "CSHL Small RNA-seq Human fetal whole eye 138T ; 139T", "file_uuids": [], "dataset_type": "experiment", "experiment_control_uuids": [], "replicate_uuids": ["46b9107d-3530-490e-c712-e913943ceadf", "f7d9d2dd-bd3b-44ec-ac96-e1bfc8b90bc6"], "dbxref": {"geo": [], "encode2": []}, "@id": "/experiments/640e1eb1-7eea-49d3-8e63-a92806a9687e/", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "dataset_accession": "ENCSR000AGI", "document_list": [], "controls": [], "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "@type": ["experiment", "item"]}], "biosamples": [{"accession": "ENCBS086RNA", "product_url": "http://www.biochain.com/index.php/catalogsearch/advanced/result/?catalog_number=R1244108-10", "starting_amount": "Unknown", "biosample_description": "Human fetal whole eye", "donor_uuid": "f6277156-eb8e-473d-95eb-761ab6860793", "lot_id": "B608426", "culture_harvest_date": "N/A", "related_biosample_uuid": "", "documents": [{"lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "document_type": "data sheet", "document": {"download": "B608426.pdf", "href": "@@download/document/B608426.pdf", "type": "application/pdf"}, "description": "BioChain Human fetal whole eye B608426 data sheet", "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "document_label": "BC_B608426", "test": "test", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "@id": "/documents/e40f45be-9294-42b1-8e2b-66783ea21426/", "@type": ["biosample_document", "item"], "document_name": "Eye_BC_B608426_data_sheet"}], "biosample_type": "Tissue", "passage_number": "N/A", "note": "Purchased as total RNA", "source": {"source_name": "BioChain Institute Inc", "url": "http://www.biochain.com", "actions": [{"profile": "/profiles/source.json", "href": "", "method": "POST", "name": "edit", "title": "Edit"}], "alias": "BioChain", "test": "test", "@id": "/sources/8599bb19-6c08-4610-9ebf-125c5346461e/", "@type": ["source", "item"]}, "dr. thomas gingeras at cold spring harbor laboratory_dbxref_list": "139T", "test": "test", "donor": {"strain background": "N/A", "organism": {"scientific_name": "Homo sapiens", "taxon_id": 9606, "organism_name": "human", "@id": "/organisms/7745b647-ff15-4ff3-9ced-b897d4e2983c/", "@type": ["organism", "item"]}, "age": "20 week fetal", "alias_source_uuids": ["8599bb19-6c08-4610-9ebf-125c5346461e"], "alias_list": ["F32"], "sex": "Female", "health_status": "Normal", "donor_id": "ENCDO064AAA", "organism_uuid": "7745b647-ff15-4ff3-9ced-b897d4e2983c", "test": "test", "@id": "/donors/f6277156-eb8e-473d-95eb-761ab6860793/", "@type": ["donor", "item"], "ethnicity": "Unknown"}, "lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "treatment_uuids": [], "constructs": [], "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "treatments": [], "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "date_obtained": "Unknown", "dbxref": {"dr. richard myers at hudson alpha": [], "dr. thomas gingeras at cold spring harbor laboratory": ["139T"], "encode2": []}, "@id": "/biosamples/af037bca-053a-4c09-8a47-be232222df45/", "document_uuids": ["e40f45be-9294-42b1-8e2b-66783ea21426"], "culture_start_date": "N/A", "product_id": "R1244108-10-D02", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "biosample_term_id": "UBERON:0000019", "construct_uuids": [], "biosample_term_name": "camera-type eye", "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "source_uuid": "8599bb19-6c08-4610-9ebf-125c5346461e", "@type": ["biosample", "item"], "related_biosample_accession": ""}, {"accession": "ENCBS085RNA", "product_url": "http://www.biochain.com/index.php/catalogsearch/advanced/result/?catalog_number=R1244108-10", "starting_amount": "Unknown", "biosample_description": "Human fetal whole eye", "donor_uuid": "998ef4ba-5aca-4dd0-9ef2-fa030628a115", "lot_id": "B608427", "culture_harvest_date": "N/A", "related_biosample_uuid": "", "documents": [{"lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "document_type": "data sheet", "document": {"download": "B608427.pdf", "href": "@@download/document/B608427.pdf", "type": "application/pdf"}, "description": "BioChain Human fetal whole eye B608427 data sheet", "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "document_label": "BC_B608427", "test": "test", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "@id": "/documents/243a7e03-3b82-478f-aeba-639e9453910d/", "@type": ["biosample_document", "item"], "document_name": "Eye_BC_B608427_data_sheet"}], "biosample_type": "Tissue", "passage_number": "N/A", "note": "Purchased as total RNA", "source": {"source_name": "BioChain Institute Inc", "url": "http://www.biochain.com", "actions": [{"profile": "/profiles/source.json", "href": "", "method": "POST", "name": "edit", "title": "Edit"}], "alias": "BioChain", "test": "test", "@id": "/sources/8599bb19-6c08-4610-9ebf-125c5346461e/", "@type": ["source", "item"]}, "dr. thomas gingeras at cold spring harbor laboratory_dbxref_list": "138T", "test": "test", "donor": {"strain background": "N/A", "organism": {"scientific_name": "Homo sapiens", "taxon_id": 9606, "organism_name": "human", "@id": "/organisms/7745b647-ff15-4ff3-9ced-b897d4e2983c/", "@type": ["organism", "item"]}, "age": "24 week fetal", "alias_source_uuids": ["8599bb19-6c08-4610-9ebf-125c5346461e"], "alias_list": ["F63"], "sex": "Female", "health_status": "Normal", "donor_id": "ENCDO067AAA", "organism_uuid": "7745b647-ff15-4ff3-9ced-b897d4e2983c", "test": "test", "@id": "/donors/998ef4ba-5aca-4dd0-9ef2-fa030628a115/", "@type": ["donor", "item"], "ethnicity": "Unknown"}, "lab_uuid": "dfc72c8c-d45c-4acd-979b-49fc93cf3c62", "treatment_uuids": [], "constructs": [], "award_uuid": "90be62e4-0757-4097-b5cf-2e6a20240a6f", "lab": {"award_number_list": ["U54HG004557", "U54HG007004"], "fax": "516-422-4109", "institute_name": "Cold Spring Harbor Laboratory", "name": "Thomas Gingeras, CSHL", "city": "Woodbury", "phone2": null, "address1": "500 Sunnyside Blvd", "address2": null, "phone1": "516-422-4105", "award_uuids": ["c68bafc1-dda3-41a9-b889-ec756f0368e1", "90be62e4-0757-4097-b5cf-2e6a20240a6f"], "url": null, "state": "NY", "postal_code": "11797", "institute_label": "CSHL", "test": "test", "awards": ["/awards/c68bafc1-dda3-41a9-b889-ec756f0368e1", "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f"], "pi_name": null, "country": "USA", "@id": "/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62/", "@type": ["lab", "item"], "lab_name": null}, "treatments": [], "award": {"award_name": null, "description": "The overall goal of this project is to generate fine-structure RNA maps in human and mouse (C57BL/6NJ) tissues and primary cell lines using a variety of high-throughput sequencing platforms, to evaluate the biological importance of novel transcripts by determining if evidence of their translated products can be identified. From each sample analyzed, we propose to isolate long (&gt;200 nucleotides) and short (&lt; 200 nucleotides) RNA in biological duplicate. Illumina-based maps for these samples will initially be generated using (1) RNA sequencing (-seq) of ribosomal (r-)RNA depleted long total RNA. (2) RNA-seq of tobacco acid pyrophosphatase (TAP) pre-treated short RNA (3) Pair-end Cap Analysis of Gene Expression (PE-CAGE) of total RNA. Additionally, for a subset of primary cell lines we will generate the above libraries from nuclear and cytoplasmic subcellular fractions. Long RNA-Seq data will be distilled down into functional elements consisting of splice junctions, polyadenylatio sites and de novo genes and transcripts. The short RNA data will be distilled into contigs representing the 5' ends of short RNAs up to the read length. PE-CAGE data will be analyzed to form clusters representing the 5' ends of transcripts linked to a tag internal to the transcript body. Importantly, each element will be assessed for reproducibility using a nonparametric Irreproducible Detection Rate (nplDR) script. Collectively, these data will allow for the detection of novel transcribed regions and supportive information as to the location of promoter regions and subcellular residence of transcripts. In aggregate, these data will be used to generate models of both noncoding and protein coding transcripts and to distinguish isoforms at complex loci necessary to obtain a comprehensive view of mammalian transcriptomes. For a subset of these samples we will simultaneously collect the genome sequence of the human donors to provide a reference map that will be used to map the RNA data against and derive information concerning allele-specific expression and RNA editing. Unannotated transcript models will be tested using long-range (PacBio/454) sequencing. Lastly, proteogenomic analysis will be done and the results compared against the unannotated transcripts.", "end_date": "2016-0731", "pi_first_name": "Thomas Raymond", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8402436", "award_end_date": null, "award_start_date": null, "number": "U54HG007004", "project": "ENCODE3", "award_number": null, "start_date": "2012-0921", "test": "test", "award_url": null, "pi_last_name": "Gingeras", "@id": "/awards/90be62e4-0757-4097-b5cf-2e6a20240a6f/", "@type": ["award", "item"], "name": "LANDSCAPE OF TRANSCRIPTION IN HUMAN AND MOUSE"}, "date_obtained": "Unknown", "dbxref": {"dr. richard myers at hudson alpha": [], "dr. thomas gingeras at cold spring harbor laboratory": ["138T"], "encode2": []}, "@id": "/biosamples/12cc1219-9c09-400c-9a9a-8d90953cf2b7/", "document_uuids": ["243a7e03-3b82-478f-aeba-639e9453910d"], "culture_start_date": "N/A", "product_id": "R1244108-10-D02", "submitter_uuid": "0abbd494-b852-433c-b360-93996f679dae", "biosample_term_id": "UBERON:0000019", "construct_uuids": [], "biosample_term_name": "camera-type eye", "project": "ENCODE3", "submitter": {"lab_uuids": ["dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "fax": "Unknown", "last_name": "Davis", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "Unknown", "first_name": "Carrie", "@type": ["user", "item"], "labs": ["/labs/dfc72c8c-d45c-4acd-979b-49fc93cf3c62"], "lab_name_list": ["Thomas Gingeras, CSHL"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/0abbd494-b852-433c-b360-93996f679dae/", "email": "davisc@cshl.edu", "job_title": "Project Manager"}, "source_uuid": "8599bb19-6c08-4610-9ebf-125c5346461e", "@type": ["biosample", "item"], "related_biosample_accession": ""}, {"encode2_dbxref_list": "GM13977", "accession": "ENCBS221AAA", "product_url": "http://ccr.coriell.org/Sections/Search/Sample_Detail.aspx?Ref=GM13977&amp;PgId=166", "starting_amount": "Unknown", "biosample_description": "lymphoblastoid cell line, Clinically affected; microcephaly; low frontal hairline; synophris; penciled arched eyebrows; short nose; crescent shaped mouth; hirsutism; micromelia; short thumbs; mental retardation; clinically normal monozygotic twin sister is GM13976", "donor_uuid": "d90607d9-a5fa-4988-8cbf-c7fe8fcd0998", "lot_id": "Unknown", "culture_harvest_date": "Unknown", "related_biosample_uuid": "", "documents": [{"lab_uuid": "c03602fb-aab1-44fd-8558-865964472ebe", "document_type": "unclassified document", "document": {"download": "GM13977_Crawford_protocol.pdf", "href": "@@download/document/GM13977_Crawford_protocol.pdf", "type": "application/pdf"}, "award_uuid": "a7ba7c03-d93f-4503-8116-63db88a1390c", "award": {"award_name": null, "description": "The goal of this proposal is to identify at high resolution all active gene regulatory elements in the human genome among cell types representative of most human tissues. We will accomplish this goal by identifying regions of open chromatin with two independent and complementary methods: DNasel hypersensitivity and Formaldehyde-Assisted Isolation of Regulatory Elements (FAIRE), combined with single-nucleosome mapping and chromatin immunoprecipitation (ChIP) for selected regulatory factors. The immediate benefit of success will be a high-quality public atlas of the human DNA regulatory elements that are likely to be active in each cell type. Identification of open chromatin regions has been one of the most accurate and robust methods to identify functional promoters, enhancers, silencers, insulators, and locus control regions in mammalian cells. A principal advantage of an open chromatin approach is that all potential sites in the genome are simultaneously assayed in an unbiased manner. DNasel hypersensitivity and FAIRE interrogate chromatin by entirely different underlying mechanisms, and therefore represent independent, cross-validating assessments of chromatin state. In addition, we will perform single-nucleosome mapping in selected cell types as an independent direct biochemical verification of open chromatin regions. For a selected subset of cell-types, we will further annotate open chromatin regions with respect to their biological activity by determining the binding location of proteins that mark transcription start sites, transcriptional units, insulators, or have broad regulatory function. The Aims of this proposal are 1) to identify all regions of open chromatin in 40 cell types by DNasel and FAIRE analyses, and 2) determine the biological function of open chromatin regions by ChIP and nucleosome mapping. To accomplish these Aims, we have assembled a team of five researchers who are leaders in their respective fields and who have contributed significantly to the ENCODE pilot project. Furthermore, we have developed a streamlined pipeline, which has been used to generate high quality whole-genome data from a number of cell types.", "end_date": "2013-0630", "pi_first_name": "Gregory E", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8320050", "award_end_date": null, "award_start_date": null, "number": "U54HG004563", "project": "ENCODE2", "award_number": null, "start_date": "2007-0927", "test": "test", "award_url": null, "pi_last_name": "Crawford", "@id": "/awards/a7ba7c03-d93f-4503-8116-63db88a1390c/", "@type": ["award", "item"], "name": "COMPREHENSIVE IDENTIFICATION OF ACTIVE FUNCTIONAL ELEMENTS IN HUMAN CHROMATIN"}, "project": "ENCODE2", "lab": {"award_number_list": ["U54HG004563"], "fax": "Unknown", "institute_name": "Duke University", "name": "Greg Crawford, Duke", "city": "Durham", "phone2": null, "address1": "CIEMAS, 2363B", "address2": "101 Science Dr., Box 3382", "phone1": "919-684-8196", "award_uuids": ["a7ba7c03-d93f-4503-8116-63db88a1390c"], "url": null, "state": "NC", "postal_code": "27708", "institute_label": "Duke", "test": "test", "awards": ["/awards/a7ba7c03-d93f-4503-8116-63db88a1390c"], "pi_name": null, "country": "USA", "@id": "/labs/c03602fb-aab1-44fd-8558-865964472ebe/", "@type": ["lab", "item"], "lab_name": null}, "document_label": "GM13977_Crawford_protocol", "submitter": {"lab_uuids": ["c03602fb-aab1-44fd-8558-865964472ebe"], "fax": "Unknown", "last_name": "Crawford", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "919-684-8196", "first_name": "Greg", "@type": ["user", "item"], "labs": ["/labs/c03602fb-aab1-44fd-8558-865964472ebe"], "lab_name_list": ["Greg Crawford, Duke"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/e74547f3-1676-43de-bcfc-9b946f0024c9/", "email": "greg.crawford@duke.edu", "job_title": "PI"}, "submitter_uuid": "e74547f3-1676-43de-bcfc-9b946f0024c9", "@id": "/documents/8ea64838-74d8-4980-a6f4-594e72fd9b1c/", "@type": ["biosample_document", "item"], "document_name": "GM13977_Crawford_protocol"}], "biosample_type": "Immortalized cell line", "passage_number": "Unknown", "source": {"source_name": "Coriell Institute for Medical Research", "url": "http://www.coriell.org/", "actions": [{"profile": "/profiles/source.json", "href": "", "method": "POST", "name": "edit", "title": "Edit"}], "alias": "Coriell", "test": "test", "@id": "/sources/409d724e-067e-4744-8d5b-c8f806136d80/", "@type": ["source", "item"]}, "donor": {"strain background": "N/A", "organism": {"scientific_name": "Homo sapiens", "taxon_id": 9606, "organism_name": "human", "@id": "/organisms/7745b647-ff15-4ff3-9ced-b897d4e2983c/", "@type": ["organism", "item"]}, "age": "Unknown", "alias_source_uuids": [], "alias_list": [], "sex": "Unknown", "health_status": "Unknown", "donor_id": "ENCDO000HUM", "organism_uuid": "7745b647-ff15-4ff3-9ced-b897d4e2983c", "test": "test", "@id": "/donors/d90607d9-a5fa-4988-8cbf-c7fe8fcd0998/", "@type": ["donor", "item"], "ethnicity": "Unknown"}, "lab_uuid": "c03602fb-aab1-44fd-8558-865964472ebe", "treatment_uuids": [], "constructs": [], "award_uuid": "a7ba7c03-d93f-4503-8116-63db88a1390c", "lab": {"award_number_list": ["U54HG004563"], "fax": "Unknown", "institute_name": "Duke University", "name": "Greg Crawford, Duke", "city": "Durham", "phone2": null, "address1": "CIEMAS, 2363B", "address2": "101 Science Dr., Box 3382", "phone1": "919-684-8196", "award_uuids": ["a7ba7c03-d93f-4503-8116-63db88a1390c"], "url": null, "state": "NC", "postal_code": "27708", "institute_label": "Duke", "test": "test", "awards": ["/awards/a7ba7c03-d93f-4503-8116-63db88a1390c"], "pi_name": null, "country": "USA", "@id": "/labs/c03602fb-aab1-44fd-8558-865964472ebe/", "@type": ["lab", "item"], "lab_name": null}, "treatments": [], "award": {"award_name": null, "description": "The goal of this proposal is to identify at high resolution all active gene regulatory elements in the human genome among cell types representative of most human tissues. We will accomplish this goal by identifying regions of open chromatin with two independent and complementary methods: DNasel hypersensitivity and Formaldehyde-Assisted Isolation of Regulatory Elements (FAIRE), combined with single-nucleosome mapping and chromatin immunoprecipitation (ChIP) for selected regulatory factors. The immediate benefit of success will be a high-quality public atlas of the human DNA regulatory elements that are likely to be active in each cell type. Identification of open chromatin regions has been one of the most accurate and robust methods to identify functional promoters, enhancers, silencers, insulators, and locus control regions in mammalian cells. A principal advantage of an open chromatin approach is that all potential sites in the genome are simultaneously assayed in an unbiased manner. DNasel hypersensitivity and FAIRE interrogate chromatin by entirely different underlying mechanisms, and therefore represent independent, cross-validating assessments of chromatin state. In addition, we will perform single-nucleosome mapping in selected cell types as an independent direct biochemical verification of open chromatin regions. For a selected subset of cell-types, we will further annotate open chromatin regions with respect to their biological activity by determining the binding location of proteins that mark transcription start sites, transcriptional units, insulators, or have broad regulatory function. The Aims of this proposal are 1) to identify all regions of open chromatin in 40 cell types by DNasel and FAIRE analyses, and 2) determine the biological function of open chromatin regions by ChIP and nucleosome mapping. To accomplish these Aims, we have assembled a team of five researchers who are leaders in their respective fields and who have contributed significantly to the ENCODE pilot project. Furthermore, we have developed a streamlined pipeline, which has been used to generate high quality whole-genome data from a number of cell types.", "end_date": "2013-0630", "pi_first_name": "Gregory E", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8320050", "award_end_date": null, "award_start_date": null, "number": "U54HG004563", "project": "ENCODE2", "award_number": null, "start_date": "2007-0927", "test": "test", "award_url": null, "pi_last_name": "Crawford", "@id": "/awards/a7ba7c03-d93f-4503-8116-63db88a1390c/", "@type": ["award", "item"], "name": "COMPREHENSIVE IDENTIFICATION OF ACTIVE FUNCTIONAL ELEMENTS IN HUMAN CHROMATIN"}, "dbxref": {"dr. richard myers at hudson alpha": [], "dr. thomas gingeras at cold spring harbor laboratory": [], "encode2": ["GM13977"]}, "@id": "/biosamples/8845d9f8-3f5b-448e-a59d-02fe1bcf0b18/", "document_uuids": ["8ea64838-74d8-4980-a6f4-594e72fd9b1c"], "culture_start_date": "Unknown", "product_id": "GM13977", "submitter_uuid": "e74547f3-1676-43de-bcfc-9b946f0024c9", "biosample_term_id": "BTO:0002062", "construct_uuids": [], "biosample_term_name": "B-lymphoblastoid cell line", "project": "ENCODE2", "submitter": {"lab_uuids": ["c03602fb-aab1-44fd-8558-865964472ebe"], "fax": "Unknown", "last_name": "Crawford", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "919-684-8196", "first_name": "Greg", "@type": ["user", "item"], "labs": ["/labs/c03602fb-aab1-44fd-8558-865964472ebe"], "lab_name_list": ["Greg Crawford, Duke"], "skype": "Unknown", "test": "test", "timezone": "UTC-5", "@id": "/users/e74547f3-1676-43de-bcfc-9b946f0024c9/", "email": "greg.crawford@duke.edu", "job_title": "PI"}, "source_uuid": "409d724e-067e-4744-8d5b-c8f806136d80", "@type": ["biosample", "item"], "related_biosample_accession": ""}], "targets": [{"lab_uuid": "fb0af3d0-3a4c-4e96-b67a-f273fe527b04", "uniprot_dbxref_list": "P56545", "award_uuid": "1a4d6443-8e29-4b4a-99dd-f93e72d42418", "award": {"award_name": null, "description": "We propose establishing the UCSC ENCODE Data Coordination Center to collect, organize, store, and provide access to data from the ENCODE project and other related projects, based on the well-established UCSC Genome Browser. The UCSC browser is a robust, flexible platform that integrates a variety of features that make accessing genomic data both convenient and information-rich. As such, it has attained a high level of acceptance throughout the biology, bioinformatics, and biomedical research communities. During the pilot phase of the ENCODE project, involving 1% of the genome, UCSC has been the repository of sequence-based data. Through participation in the pilot phase, we have established cost-effective, high-throughput approaches for incorporating and displaying ENCODE data, and we have developed an effective interface to the consortium for uploading data files and methods documentation to our FTP site. Through ongoing expansion and improvement of our ENCODE browser, we have laid the groundwork for data management in the whole-genome phase of the project. UCSC's specific aims for participation as the ENCODE DCC are as follows: 1) coordinate closely with the ENCODE data providers to collect, store, and manage ENCODE sequence-based functional element data and related metadata; 2) provide and maintain a long-term storage system for data from the ENCODE Consortium and related genomic projects; 3) provide the ENCODE and public research communities with freely-available mechanisms to access, search, and analyze the ENCODE data; 4) employ good software engineering practices to ensure that the ENCODE data are of the highest possible quality; and 5) foster collaboration with the biomedical community. Fundamental to understanding human biology and disease is the ability to identify and locate the functional elements in the human genome sequence. Both the effective presentation of experimental data and the ability to analyze sequence data using computational methods will help elucidate genomic function. Because the UCSC Genome Browser serves both of these purposes, it offers the ideal platform for ENCODE data.", "end_date": "2012-1231", "pi_first_name": "W James", "url": "http://projectreporter.nih.gov/project_info_details.cfm?aid=8332919", "award_end_date": null, "award_start_date": null, "number": "U41HG004568", "project": "ENCODE2", "award_number": null, "start_date": "2007-0928", "test": "test", "award_url": null, "pi_last_name": "Kent", "@id": "/awards/1a4d6443-8e29-4b4a-99dd-f93e72d42418/", "@type": ["award", "item"], "name": "THE UCSC ENCODE DATA COORDINATION CENTER"}, "target_gene_name": "CTBP2", "is_current": true, "lab": {"award_number_list": ["U41HG004568"], "fax": "Unknown", "institute_name": "University of California at Santa Cruz", "name": "Jim Kent, UCSC", "city": "Santa Cruz", "phone2": null, "address1": "Baskin School of Engineering", "address2": "1156 High St.; Mail Stop CBSE/ITI", "phone1": "831-459-1401", "award_uuids": ["1a4d6443-8e29-4b4a-99dd-f93e72d42418"], "url": null, "state": "CA", "postal_code": "95064", "institute_label": "UCSC", "test": "test", "awards": ["/awards/1a4d6443-8e29-4b4a-99dd-f93e72d42418"], "pi_name": null, "country": "USA", "@id": "/labs/fb0af3d0-3a4c-4e96-b67a-f273fe527b04/", "@type": ["lab", "item"], "lab_name": null}, "organism_uuid": "7745b647-ff15-4ff3-9ced-b897d4e2983c", "ensembl_dbxref": "ENSG00000175029", "target_class": "Unknown", "submitter_uuid": "746eef27-d857-4b38-a469-cac93fb02164", "@id": "/targets/0369858d-7f87-42c4-84ff-7a352e825ffa/", "dbxref": {"geneid": ["1488"], "uniprot": ["P56545"], "hgnc": ["ribeye"]}, "target_label": "CTBP2", "project": "ENCODE2", "geneid_dbxref_list": 1488, "submitter": {"lab_uuids": ["fb0af3d0-3a4c-4e96-b67a-f273fe527b04", "cfb789b8-46f3-4d59-a2b3-adc39e7df93a"], "fax": "Unknown", "last_name": "Kent", "phone2": "Unknown", "google_chat": "Unknown", "phone1": "831-459-1401", "first_name": "Jim", "@type": ["user", "item"], "labs": ["/labs/fb0af3d0-3a4c-4e96-b67a-f273fe527b04", "/labs/cfb789b8-46f3-4d59-a2b3-adc39e7df93a"], "lab_name_list": ["Jim Kent, UCSC", "Michael Cherry, Stanford"], "skype": "Unknown", "test": "test", "timezone": "UTC-8", "@id": "/users/746eef27-d857-4b38-a469-cac93fb02164/", "email": "kent@soe.ucsc.edu", "job_title": "PI"}, "date_created": "2012-09-30", "hgnc_dbxref_list": "ribeye", "organism": {"scientific_name": "Homo sapiens", "taxon_id": 9606, "organism_name": "human", "@id": "/organisms/7745b647-ff15-4ff3-9ced-b897d4e2983c/", "@type": ["organism", "item"]}, "@type": ["target", "item"]}]}, "@id": "/search/", "@type": ["search"]}</script>
        <div id="slot-application"><div id="application" class="done" data-reactid=".r[5q14e]"><div id="layout" data-reactid=".r[5q14e].[0]"><div id="navbar" class="navbar navbar-fixed-top navbar-inverse" data-reactid=".r[5q14e].[0].[0]"><div class="navbar-inner" data-reactid=".r[5q14e].[0].[0].[0]"><div class="container" data-reactid=".r[5q14e].[0].[0].[0].[0]"><a class="btn btn-navbar" href="" data-toggle="collapse" data-target=".nav-collapse" data-reactid=".r[5q14e].[0].[0].[0].[0].[0]"><span class="icon-bar" data-reactid=".r[5q14e].[0].[0].[0].[0].[0].[0]"></span><span class="icon-bar" data-reactid=".r[5q14e].[0].[0].[0].[0].[0].[1]"></span><span class="icon-bar" data-reactid=".r[5q14e].[0].[0].[0].[0].[0].[2]"></span></a><a class="brand" href="http://staging.encodedcc.org/" data-reactid=".r[5q14e].[0].[0].[0].[0].[1]">ENCODE</a><div class="nav-collapse collapse" data-reactid=".r[5q14e].[0].[0].[0].[0].[2]"><ul id="global-sections" class="nav" data-reactid=".r[5q14e].[0].[0].[0].[0].[2].[0]"><li class="" data-reactid=".r[5q14e].[0].[0].[0].[0].[2].[0].[antibodies]"><a href="http://staging.encodedcc.org/antibodies/" data-reactid=".r[5q14e].[0].[0].[0].[0].[2].[0].[antibodies].[0]">Antibodies</a></li><li class="" data-reactid=".r[5q14e].[0].[0].[0].[0].[2].[0].[biosamples]"><a href="http://staging.encodedcc.org/biosamples/" data-reactid=".r[5q14e].[0].[0].[0].[0].[2].[0].[biosamples].[0]">Biosamples</a></li><li class="" data-reactid=".r[5q14e].[0].[0].[0].[0].[2].[0].[experiments]"><a href="http://staging.encodedcc.org/experiments/" data-reactid=".r[5q14e].[0].[0].[0].[0].[2].[0].[experiments].[0]">Experiments</a></li><li class="" data-reactid=".r[5q14e].[0].[0].[0].[0].[2].[0].[targets]"><a href="http://staging.encodedcc.org/targets/" data-reactid=".r[5q14e].[0].[0].[0].[0].[2].[0].[targets].[0]">Targets</a></li></ul><ul id="user-actions" class="nav pull-right" data-reactid=".r[5q14e].[0].[0].[0].[0].[2].[1]"><li data-reactid=".r[5q14e].[0].[0].[0].[0].[2].[1].[0]"><a href="" data-trigger="login" data-id="signin" data-reactid=".r[5q14e].[0].[0].[0].[0].[2].[1].[0].[0]">Sign in</a></li></ul></div></div></div></div><div id="content" class="container" data-reactid=".r[5q14e].[0].[1]"><div data-reactid=".r[5q14e].[0].[1].[0]"><form class="input-prepend" data-reactid=".r[5q14e].[0].[1].[0].[0]"><span class="add-on" data-reactid=".r[5q14e].[0].[1].[0].[0].[0]"><i class="icon-search" data-reactid=".r[5q14e].[0].[1].[0].[0].[0].[0]">Search</i></span><input class="search-query" placeholder="Search ENCODE" name="searchTerm" data-reactid=".r[5q14e].[0].[1].[0].[0].[1]" type="text"></form>
       
<!-- ALL STYLES FOR FACET STUFF IS BELOW --> 
<style>
	.box {
		padding: 10px;
		border: 1px #ccc solid;
		border-radius: 3px;
		overflow: visible;
	}
	.facet { margin-right: 20px; }
	.facet ul.facet-list { 
		list-style: none; 
		margin: 0; 
		font-weight: bold;
	}
	.facet h4 {
		margin: 0 0 4px 0;
	}
	.facet input { vertical-align: middle; }
	.facet input[type="radio"] { margin-top:-1px; }
	.searchResults {
		list-style: none;
		margin: 10px 0;
	}
	.result {
		margin: 0 0 25px 0; 
	}
	.result h3 {
		margin: 0;
		padding: 0;
		font-size: 1.2em;
		line-height: 1.2em;
	}
	.resultDesc {
		
	}
	.resultMeta {
	
	}
	.facet-list li {
		padding: 2px 10px;
		margin: 0 0 6px 0;
		border-radius: 10px;
		position: relative;
	}
	.facet li.selected {
		font-weight: bold;
		background-color: #ededed;
	}
	.facet li.selected:after { left: 99.5%; border: solid transparent; content: " "; height: 0; width: 0; position: absolute; pointer-events: none; } 
	.facet li.selected:after { border-color: rgba(191, 191, 191, 0); border-left-color: #ededed; border-width: 7px; top: 50%; margin-top: -7px; }
	
</style>

<div class="panel data-display" data-reactid=".r[5q14e].[0].[1].[0].[1]">
	<div class="row">
        <div class="span3" id="facets">
        	<h4>Filter results</h4>
        	<section class="facet wel box">
        		<h4>Data type</h4>
        		<ul class="facet-list">
        			<li class="selected">All (863)</li>
        			<li><a href="">Antibodies (432)</a></li>
        			<li><a href="">Biosamples (219)</a></li>
        			<li><a href="">Experiments (43)</a></li>
        			<li><a href="">Targets (12)</a></li>
        		</ul>
        	</section>
        </div><!-- #facets -->
        
        <div class="span8">
        	<ul class="searchResults">
        		<li class="result">
        			<h3>Result Title</h3>
        			<div class="resultDesc">
        				Some descriptive stuff.
        			</div>
        			<div class="resultMeta">
        				Meta data - species - fluff - yazoo
        			</div>
        		</li>
        		<li class="result">
        			<h3>Result Title</h3>
        			<div class="resultDesc">
        				Some descriptive stuff.
        			</div>
        			<div class="resultMeta">
        				Meta data - species - fluff - yazoo
        			</div>
        		</li>
        		<li class="result">
        			<h3>Result Title</h3>
        			<div class="resultDesc">
        				Some descriptive stuff.
        			</div>
        			<div class="resultMeta">
        				Meta data - species - fluff - yazoo
        			</div>
        		</li>
        		<li class="result">
        			<h3>Result Title</h3>
        			<div class="resultDesc">
        				Some descriptive stuff.
        			</div>
        			<div class="resultMeta">
        				Meta data - species - fluff - yazoo
        			</div>
        		</li>
        		
        	</ul>
        </div>
	</div>        
</div>
        
        
        
        
        
        
        
        
        
        
        
        </div></div><div id="layout-footer" data-reactid=".r[5q14e].[0].[3]"></div></div><footer id="page-footer" data-reactid=".r[5q14e].[1]"><div id="footer-top" data-reactid=".r[5q14e].[1].[0]"><div class="container" data-reactid=".r[5q14e].[1].[0].[0]"><a href="http://www.stanford.edu/" data-reactid=".r[5q14e].[1].[0].[0].[0]"><img src="SearchENCODE_files/su-logo.png" alt="Stanford University" id="su-logo" data-reactid=".r[5q14e].[1].[0].[0].[0].[0]"></a><a href="http://med.stanford.edu/" data-reactid=".r[5q14e].[1].[0].[0].[1]"><img src="SearchENCODE_files/som-logo.png" alt="Stanford University School of Medicine" id="som-logo" data-reactid=".r[5q14e].[1].[0].[0].[1].[0]"></a><div id="footer-links" data-reactid=".r[5q14e].[1].[0].[0].[2]"><ul data-reactid=".r[5q14e].[1].[0].[0].[2].[0]"><li data-reactid=".r[5q14e].[1].[0].[0].[2].[0].[0]"><a href="http://staging.encodedcc.org/" data-reactid=".r[5q14e].[1].[0].[0].[2].[0].[0].[0]">ENCODE</a></li><li data-reactid=".r[5q14e].[1].[0].[0].[2].[0].[1]"><a href="" data-reactid=".r[5q14e].[1].[0].[0].[2].[0].[1].[0]">Contact</a></li><li data-reactid=".r[5q14e].[1].[0].[0].[2].[0].[2]"><a href="http://www.stanford.edu/site/terms.html" data-reactid=".r[5q14e].[1].[0].[0].[2].[0].[2].[0]">Terms of Use</a></li></ul></div></div></div><div id="footer-bottom" data-reactid=".r[5q14e].[1].[1]"><div class="container" data-reactid=".r[5q14e].[1].[1].[0]"><span data-reactid=".r[5q14e].[1].[1].[0].[0]"> 2013. </span><a href="http://www.stanford.edu/" data-reactid=".r[5q14e].[1].[1].[0].[1]">Stanford University</a><span data-reactid=".r[5q14e].[1].[1].[0].[2]">. </span><a href="http://med.stanford.edu/" data-reactid=".r[5q14e].[1].[1].[0].[3]">School of Medicine</a><span data-reactid=".r[5q14e].[1].[1].[0].[4]">. </span><a href="http://genetics.stanford.edu/" data-reactid=".r[5q14e].[1].[1].[0].[5]">Department of Genetics</a><span data-reactid=".r[5q14e].[1].[1].[0].[6]">. </span></div></div></footer></div></div>
    

<iframe src="SearchENCODE_files/communication_iframe.html" style="display: none;"></iframe></body></html>